<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="fido"><title>Introduction to fido::Pibble • fido</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to fido::Pibble"><meta property="og:description" content="fido"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">fido</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction-to-fido.html">Introduction to fido::Pibble</a>
    <a class="dropdown-item" href="../articles/mitigating-pcrbias.html">Example of using Fido for measuring and mitigating PCR Bias</a>
    <a class="dropdown-item" href="../articles/non-linear-models.html">Non-linear models with fido::basset</a>
    <a class="dropdown-item" href="../articles/orthus.html">Joint Modeling  (e.g., Multiomics) with fido::Orthus</a>
    <a class="dropdown-item" href="../articles/picking_priors.html">Picking Priors</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jsilve24/fido/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to fido::Pibble</h1>
                        <h4 data-toc-skip class="author">Justin
Silverman</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jsilve24/fido/blob/HEAD/vignettes/introduction-to-fido.Rmd" class="external-link"><code>vignettes/introduction-to-fido.Rmd</code></a></small>
      <div class="d-none name"><code>introduction-to-fido.Rmd</code></div>
    </div>

    
    
<div id="an-introduction-to-fido" class="section level1">
<h1>An introduction to <em>fido</em></h1>
<p><em>fido</em> <span class="citation">(Justin D. Silverman et al.
2019)</span> is a loose acronym for <strong>“(Bayesian) Multinomial
Logistic-Normal Models”</strong>. In particular the development of
<em>fido</em> stems from the need for fast inference for time-invariant
MALLARD models<span class="citation">(Justin D. Silverman et al.
2018)</span>. <em>fido</em> is very fast! It uses closed form solutions
for model gradients and Hessians written in C++ to preform <a
href="https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation">MAP
estimation</a> in combination with parameter uncertainty estimation
using a <a
href="https://www.sumsar.net/blog/2013/11/easy-laplace-approximation/">Laplace
Approximation</a>. One of the main models in <em>fido</em> is the
function <em>pibble</em> which fits a Multinomial Logistic-Normal
<strong>Linear Regression</strong> model.</p>
<p><strong>So what is a <em>fido</em> model exactly?</strong> First let
me give the broad description from 10,000ft up: Basically its a model
for multinomial count data (e.g., each sample contains the counts of
<span class="math inline">\(D\)</span> “types of things”). Importantly,
unlike the more common Poisson count models, the multinomial models a
“competition to be counted” (i.e., cases in which counting more of one
type of thing means that I have less resources available to count other
types of things).</p>
<p>This may seem vague so let me give an example. Pretend there is a
ball pit with red, green, and blue balls. Pretend that the ball pit is
very large and I don’t know the total number of balls in the ball pit,
yet I want to say something about the relative number of red, blue, and
green balls in the pit. One way I may choose to measure the ball pit is
by grabbing an armful of balls and counting the number of balls of each
color (e.g., in one armful I may collect 5 red, 3 blue, and 6 green). My
arms can only contain so many balls (in this example about 14) and so if
I were to have (randomly) gotten another green ball in my armful (making
7 total) I would likely not have been able to measure one of the red or
blue balls; hence the “competition to be counted”. It turns out that
this type of sampling occurs all the time in many situations (Wikipedia
has an example with <a
href="https://en.Wikipedia.org/wiki/Multinomial_distribution#Example">political
polling</a>). Perhaps one of the most notable examples of this type of
count data occurs with modern high-throughput sequencing studies such as
16S rRNA studies to profile microbial communities or bulk/single-cell
RNA-seq studies to study expression profiles of cells. In all cases,
transcripts are sequenced and the number of different types of
transcripts are counted. The important part is that sequencing only
samples a small portion of the total genetic material available and
leads to similar competition to be counted.</p>
<div id="the-pibble-model" class="section level2">
<h2>The <em>pibble</em> model</h2>
<p><em>Pibble</em> is one type of <em>fido</em> model. In particular its
a <em>fido</em> model for multivariate linear regression.</p>
<p>Let <span class="math inline">\(Y\)</span> denote an <span
class="math inline">\(D\times N\)</span> matrix of counts. Let us denote
the <span class="math inline">\(j\)</span>-th column of <span
class="math inline">\(Y\)</span> as <span
class="math inline">\(Y_j\)</span>. Thus each “sample” in the dataset is
a measurement of the relative amount of <span
class="math inline">\(D\)</span> “types of things”. Suppose we also have
have covariate information in the form of a <span
class="math inline">\(Q\times N\)</span> matrix <span
class="math inline">\(X\)</span>.</p>
<p>The following is the pibble model including likelihood and priors:
<span class="math display">\[
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j \right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\eta_j &amp;\sim N(\Lambda X_j, \Sigma) \\
\Lambda &amp;\sim  MN_{(D-1) \times Q}(\Theta, \Sigma, \Gamma) \\
\Sigma &amp;\sim W^{-1}(\Xi, \upsilon)
\end{align}
\]</span> Here <span class="math inline">\(MN_{(D-1) \times Q}\)</span>
denotes a <a
href="https://en.wikipedia.org/wiki/Matrix_normal_distribution">Matrix
Normal distribution</a> for a matrix <span
class="math inline">\(\Lambda\)</span> of regression coefficients of
dimension <span class="math inline">\((D-1)\times Q\)</span>.
Essentially you can think of the Matrix normal as having two covariance
matrices one describing the covariation between the rows of <span
class="math inline">\(\Lambda\)</span> (<span
class="math inline">\(\Sigma\)</span>) and another describing the
covariation of the columns of <span
class="math inline">\(\Lambda\)</span> (<span
class="math inline">\(\Gamma\)</span>). and <span
class="math inline">\(W^{-1}\)</span> refers to the <a
href="https://en.wikipedia.org/wiki/Inverse-Wishart_distribution">Inverse
Wishart distribution</a> (which is a common distribution over covariance
matrices). The line <span class="math inline">\(\pi_j =
\phi^{-1}(\eta_j)\)</span> represents a transformation between the
parameters <span class="math inline">\(\pi_j\)</span> which exist on a
simplex (e.g., <span class="math inline">\(\pi_j\)</span> must sum to 1)
and the transformed parameters <span
class="math inline">\(\eta_j\)</span> that exist in real space. In
particular we define <span class="math inline">\(\phi^{-1}\)</span> to
be the <a
href="http://www.sediment.uni-goettingen.de/staff/tolosana/extra/CoDaNutshell.pdf">inverse
additive log ratio transform</a> (which conversely implies that <span
class="math inline">\(\eta_j = ALR(\pi_j)\)</span>) also known as the
identified softmax transform (as it is more commonly known in the
Machine Learning community). While I will say more on this later in this
tutorial, one thing to know is that I have the model implemented using
the ALR transform as it is computationally simple and fast; the results
of the model can be viewed as if any number of transforms had been used
(instead of the ALR) including the isometric log-ratio transform, or the
centered log-ratio transform.</p>
<p>Before moving on, I would like to give <strong>a more intuitive
description of <em>pibble</em></strong>. Essentially the main modeling
component of <em>pibble</em> is the third equation above (<span
class="math inline">\(\eta_j \sim N(\Lambda X_j, \Sigma)\)</span>) which
is just a multivariate linear model. That is, <span
class="math inline">\(X\)</span> are your covariates (which can be
continuous, discrete, binary, etc…), and <span
class="math inline">\(\Sigma\)</span> is the covariance matrix for the
regression residuals.</p>
</div>
</div>
<div id="example-analysis-of-microbiome-data" class="section level1">
<h1>Example analysis of microbiome data</h1>
<p>This analysis is the same as that presented in the <em>fido</em>
manuscript <span class="citation">(Justin D. Silverman et al.
2019)</span>. I will reanalyze a previously published study comparing
microbial composition in the terminal ileum of subjects with Crohn’s
Disease (CD) to healthy controls <span class="citation">(Gevers et al.
2014)</span>. To do this I will fit a pibble model using CD status,
inflammation status and age as covariates (plus a constant intercept
term).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(MicrobeDS)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(fido)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">899</span>)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;RISK_CCFA&quot;</span>)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co"># drop low abundant taxa and samples</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>dat <span class="ot">&lt;-</span> RISK_CCFA <span class="sc">%&gt;%</span> </span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="fu">subset_samples</span>(disease_stat<span class="sc">!=</span><span class="st">&quot;missing&quot;</span>, </span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>                 immunosup<span class="sc">!=</span><span class="st">&quot;missing&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>  <span class="fu">subset_samples</span>(diagnosis <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;no&quot;</span>, <span class="st">&quot;CD&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>  <span class="fu">subset_samples</span>(steroids<span class="sc">==</span><span class="st">&quot;false&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>  <span class="fu">subset_samples</span>(antibiotics<span class="sc">==</span><span class="st">&quot;false&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>  <span class="fu">subset_samples</span>(biologics<span class="sc">==</span><span class="st">&quot;false&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>  <span class="fu">subset_samples</span>(biopsy_location<span class="sc">==</span><span class="st">&quot;Terminal ileum&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>  <span class="fu">tax_glom</span>(<span class="st">&quot;Family&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>  <span class="fu">prune_samples</span>(<span class="fu">sample_sums</span>(.) <span class="sc">&gt;=</span> <span class="dv">5000</span>,.) <span class="sc">%&gt;%</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(<span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">3</span>) <span class="sc">&gt;</span> <span class="fl">0.10</span><span class="sc">*</span><span class="fu">length</span>(x), <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Create Design Matrix and OTU Table</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>sample_dat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as</span>(<span class="fu">sample_data</span>(dat),<span class="st">&quot;matrix&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">age =</span> <span class="fu">as.numeric</span>(<span class="fu">as.character</span>(age)),</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>         <span class="at">diagnosis =</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(diagnosis, <span class="at">ordered =</span> <span class="cn">FALSE</span>), <span class="at">ref=</span><span class="st">&quot;no&quot;</span>), </span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>         <span class="at">disease_stat =</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(disease_stat, <span class="at">ordered =</span> <span class="cn">FALSE</span>), <span class="at">ref=</span><span class="st">&quot;non-inflamed&quot;</span>))</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">model.matrix</span>(<span class="sc">~</span>diagnosis <span class="sc">+</span> disease_stat<span class="sc">+</span>age, <span class="at">data=</span>sample_dat))</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>Y <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(dat)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Investigate X and Y look like</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>X[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt;                      1939.SKBTI.0175 1939.SKBTI047 1939.SKBTI051 1939.SKBTI063</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; (Intercept)                  1.00000       1.00000          1.00       1.00000</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; diagnosisCD                  1.00000       1.00000          1.00       1.00000</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; disease_statinflamed         0.00000       1.00000          1.00       1.00000</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; age                         15.16667      14.33333         15.75      13.58333</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt;                      1939.SKBTI072</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; (Intercept)                   1.00</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt; diagnosisCD                   1.00</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt; disease_statinflamed          1.00</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt; age                          15.75</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>Y[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt; OTU Table:          [5 taxa and 5 samples]</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt;                      taxa are rows</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt;         1939.SKBTI.0175 1939.SKBTI047 1939.SKBTI051 1939.SKBTI063 1939.SKBTI072</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt; 4442127               0             9             0            14             2</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt; 74305                 1             2            35             1             0</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt; 663573               36             1             0             2             1</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt; 2685602              10           264           211           276            83</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt; 4339819               0            37            42            70            22</span></span></code></pre></div>
<p>Next specify priors. We are going to start by specifying a prior on
the covariance between log-ratios <span
class="math inline">\(\Sigma\)</span>. I like to do this by thinking
about a prior on the covariance between taxa on the log-scale (i.e.,
between the log of their absolute abundances not the log-ratios). I will
refer to this covariance on log-absolute abundances <span
class="math inline">\(\Omega\)</span>. For example, here I will build a
prior that states that the mean of <span
class="math inline">\(\Omega\)</span> is the identity matrix <span
class="math inline">\(I_D\)</span>. From From <span
class="citation">Aitchison (1986)</span>, we know that if we assume that
the taxa have a covariance <span class="math inline">\(\Omega\)</span>
in terms of log-absolute abundance then their correlation in the <span
class="math inline">\(\text{ALR}_D\)</span> is given by <span
class="math display">\[ \Sigma = G \Omega G^T \]</span> where <span
class="math inline">\(G\)</span> is a <span class="math inline">\(D-1
\times D\)</span> matrix given by <span class="math inline">\(G =
[I_{D-1}; -1_{D-1}]\)</span> (i.e., <span
class="math inline">\(G\)</span> is the <span
class="math inline">\(\text{ALR}_D\)</span> contrast matrix).
Additionally, we know that the Inverse Wishart mode is given by <span
class="math inline">\(\frac{\Xi}{\upsilon + D}\)</span>. Finally, note
that <span class="math inline">\(\upsilon\)</span> essentially controls
our uncertainty in <span class="math inline">\(\Sigma\)</span> about
this prior mean. Here I will take <span class="math inline">\(\upsilon =
D+3\)</span>. This then gives us <span class="math inline">\(\Xi =
(\upsilon - D) GIG^T\)</span>. We scale <span
class="math inline">\(\Xi\)</span> by a factor of 1/2 to make <span
class="math inline">\(Tr(\Xi)=D-1\)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>upsilon <span class="ot">&lt;-</span> <span class="fu">ntaxa</span>(dat)<span class="sc">+</span><span class="dv">3</span> </span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>Omega <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">ntaxa</span>(dat))</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>G <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">diag</span>(<span class="fu">ntaxa</span>(dat)<span class="sc">-</span><span class="dv">1</span>), <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>Xi <span class="ot">&lt;-</span> (upsilon<span class="sc">-</span><span class="fu">ntaxa</span>(dat))<span class="sc">*</span>G<span class="sc">%*%</span>Omega<span class="sc">%*%</span><span class="fu">t</span>(G)</span></code></pre></div>
<p>Finally I specify my priors for <span
class="math inline">\(\Theta\)</span> (mean of <span
class="math inline">\(\Lambda\)</span>) and <span
class="math inline">\(\Gamma\)</span> (covariance between columns of
<span class="math inline">\(\Lambda\)</span>; i.e., covariance between
the covariates). I will center my prior for <span
class="math inline">\(\Lambda\)</span> about zero, and assume that the
covariates are independent.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>Theta <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">ntaxa</span>(dat)<span class="sc">-</span><span class="dv">1</span>, <span class="fu">nrow</span>(X))</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>Gamma <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">nrow</span>(X))</span></code></pre></div>
<p>I strongly recommend users perform prior predictive checks to make
sure their priors make sense to them. <em>fido</em> makes this easy, all
the main fitting functions (e.g., <code>pibble</code>) will
automatically sample from the prior predictive distribution if
<code>Y</code> is left as <code>NULL</code> (e.g., without data your
posterior is just your prior).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>priors <span class="ot">&lt;-</span> <span class="fu">pibble</span>(<span class="cn">NULL</span>, X, upsilon, Theta, Gamma, Xi)  </span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">print</span>(priors)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt;  pibblefit Object (Priors Only): </span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt;   Number of Samples:      250 </span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt;   Number of Categories:       49 </span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt;   Number of Covariates:       4 </span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt;   Number of Posterior Samples:    2000 </span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt;   Contains Samples of Parameters:Eta  Lambda  Sigma</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;   Coordinate System:      alr, reference category: 49</span></span></code></pre></div>
<p>The main fitting functions in the <em>fido</em> package output
special fit objects (e.g., <code>pibble</code> outputs an object of
class <code>pibblefit</code>). These fit objects are just lists with
some extra metadata that allows special method dispatch. For example, if
you call print on a <code>pibblefit</code> object you will get a nice
summary of what is in the object.</p>
<p><em>Note:</em> Currently, the function <code>pibble</code> takes
expects inputs and outputs in the “default” coordinate system; this is
simply the ALR coordinate system where the last category (49 above) is
taken as reference (this will be generalized in future versions). More
specifically for a vector <span class="math inline">\(x\)</span>
representing the proportions of categories <span
class="math inline">\(\{1, \dots, D\}\)</span> we can write <span
class="math display">\[x^* = \left( \log \frac{x_1}{x_D}, \dots, \log
\frac{x_{D-1}}{x_D}\right).\]</span> As mentioned above however, I have
designed <em>fido</em> to work with many different coordinate systems
including the ALR (with respect to any category), CLR, ILR, or
proportions. To help transform things between these coordinate systems I
have written a series of transformation functions that transform any
<code>pibblefit</code> object into a desired coordinate system.
Importantly, <code>pibblefit</code> objects keep track of what
coordinate system they are currently in so as a user you only need to
specify the coordinate system that you want to change into. Keep in mind
that covariance matrices cannot be represented in proportions and so
visualizations or summaries based on covariance matrices will be
suppressed when <code>pibblefit</code> objects are in the proportions
coordinate system. As an example, lets look at viewing a summary of the
prior for <span class="math inline">\(\Lambda\)</span> with respect to
the CLR coordinate system<a href="#fn1" class="footnote-ref"
id="fnref1"><sup>1</sup></a>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>priors <span class="ot">&lt;-</span> <span class="fu">to_clr</span>(priors)  </span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">summary</span>(priors, <span class="at">pars=</span><span class="st">&quot;Lambda&quot;</span>, <span class="at">gather_prob=</span><span class="cn">TRUE</span>, <span class="at">as_factor=</span><span class="cn">TRUE</span>, <span class="at">use_names=</span><span class="cn">TRUE</span>)  </span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; $Lambda</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt; # A tibble: 784 x 9</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt;    Parameter coord covariate         val .lower .upper .width .point .interval</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;       &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    </span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt;  1 Lambda        1         1  0.0509     -0.528  0.596    0.5 mean   qi       </span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt;  2 Lambda        1         2  0.00199    -0.567  0.575    0.5 mean   qi       </span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt;  3 Lambda        1         3 -0.0373     -0.620  0.532    0.5 mean   qi       </span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt;  4 Lambda        1         4 -0.00205    -0.554  0.549    0.5 mean   qi       </span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt;  5 Lambda        2         1  0.00991    -0.564  0.535    0.5 mean   qi       </span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt;  6 Lambda        2         2 -0.00000782 -0.572  0.580    0.5 mean   qi       </span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt;  7 Lambda        2         3 -0.0247     -0.570  0.518    0.5 mean   qi       </span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt;  8 Lambda        2         4  0.0165     -0.536  0.589    0.5 mean   qi       </span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt;  9 Lambda        3         1  0.0138     -0.584  0.620    0.5 mean   qi       </span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt; 10 Lambda        3         2 -0.00502    -0.542  0.581    0.5 mean   qi       </span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt; # ... with 774 more rows</span></span></code></pre></div>
<p>By default the <code>summary</code> function returns a list (with
possible elements <code>Lambda</code>, <code>Sigma</code>, and
<code>Eta</code>) summarizing each posterior parameter based on
quantiles and mean (e.g., p2.5 is the 0.025 percentile of the posterior
distribution). As this type of table may be hard to take in due to how
large it is, <code>pibblefit</code> objects also come with a default
plotting option for each of the parameters. Also the returned plot
objects are <code>ggplot</code> objects so normal <code>ggplot2</code>
commands work on them. Before doing that though we are going to use one
of the <code>names</code> functions for <code>pibblefit</code> objects
to provide some more specific names for the covariates (helpful when we
then plot).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">names_covariates</span>(priors) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(X)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot</span>(priors, <span class="at">par=</span><span class="st">&quot;Lambda&quot;</span>) </span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; Scale for &#39;colour&#39; is already present. Adding another scale for &#39;colour&#39;, which will</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt; replace the existing scale.</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>p <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">10</span>))  </span></code></pre></div>
<p><img src="https://github.com/jsilve24/fido/raw/master/vignettes/fido-lambda-all.png" width="50%" style="display: block; margin: auto;" /></p>
<p>This looks fairly reasonable to me. So I am going to go ahead and fit
the model with data. <code>fido</code> provides a helper method called
<code>refit</code> that we will use to avoid passing prior parameters
again.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>priors<span class="sc">$</span>Y <span class="ot">&lt;-</span> Y <span class="co"># remember pibblefit objects are just lists</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>posterior <span class="ot">&lt;-</span> <span class="fu">refit</span>(priors, <span class="at">optim_method=</span><span class="st">&quot;lbfgs&quot;</span>)</span></code></pre></div>
<p>Unlike the main <em>pibble</em> function, the <code>refit</code>
method can be called on objects in any coordinate system and all
transformations to and from the default coordinate system are handled
internally<a href="#fn2" class="footnote-ref"
id="fnref2"><sup>2</sup></a>. This is one nice thing about using the
<code>refit</code> method. That said, new objects added to the
<code>pibblefit</code> object need to be added in the proper coordinates
For example, if we wanted to replace our prior for <span
class="math inline">\(\Xi\)</span> for an object in CLR coordinates, we
would had to transform our prior for <code>Xi</code> to CLR coordinates
before adding it to the <code>priors</code> object.</p>
<p>Now I are also going to add in the taxa names to make it easier to
interpret the results.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>tax <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(dat)[,<span class="fu">c</span>(<span class="st">&quot;Class&quot;</span>, <span class="st">&quot;Family&quot;</span>)]</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>tax <span class="ot">&lt;-</span> <span class="fu">apply</span>(tax, <span class="dv">1</span>, paste, <span class="at">collapse=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">names_categories</span>(posterior) <span class="ot">&lt;-</span> tax</span></code></pre></div>
<p>Before doing anything else lets look at the posterior predictive
distribution to assess model fit. This can be accessed through the
method <code>ppc</code><a href="#fn3" class="footnote-ref"
id="fnref3"><sup>3</sup></a>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">ppc</span>(posterior) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">coord_cartesian</span>(<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">30000</span>))</span></code></pre></div>
<p><img src="https://github.com/jsilve24/fido/raw/master/vignettes/fido-ppc.png" width="50%" style="display: block; margin: auto;" /></p>
<p>There are a few things to note about this plot. First, when zoomed
out like this it looks it is hard to make much of it. This is a fairly
large dataset we are analyzing and its hard to view an uncertainty
interval; in this case its plotting the median and 95% confidence
interval in grey and black and the observed counts in green.
<em>fido</em> also has a simpler function that summarizes the posterior
predictive check.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">ppc_summary</span>(posterior)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co">#&gt; Proportions of Observations within 95% Credible Interval: 0.9898776</span></span></code></pre></div>
<p>Here we see that the model appears to be fitting well (at least based
on the posterior predictive check) and that only about 1.5% of
observations fall outside of the 95% posterior predictive density (this
is good).</p>
<p>Some readers will look at the above <code>ppc</code> plots and think
“looks like over-fitting”. However, note that there are two ways of
using <code>ppc</code>. One is to predict the counts based on the
samples of <span class="math inline">\(\eta\)</span> (Eta; as we did
above); the other is to predict “from scratch” that is to predict
starting form the posterior samples of <span
class="math inline">\(\Lambda\)</span> (Lambda) then sampling <span
class="math inline">\(\eta\)</span> and only then sampling <span
class="math inline">\(Y\)</span>. This later functionality can be
accessed by also passing the parameters <code>from_scratch=TRUE</code>
to the <code>ppc</code> function. Note: these two posterior predictive
checks have different meanings, one is not better than the other.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">ppc</span>(posterior, <span class="at">from_scratch=</span><span class="cn">TRUE</span>) <span class="sc">+</span>ggplot2<span class="sc">::</span><span class="fu">coord_cartesian</span>(<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">30000</span>))</span></code></pre></div>
<p><img src="https://github.com/jsilve24/fido/raw/master/vignettes/fido-ppc-scratch.png" width="50%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">ppc_summary</span>(posterior, <span class="at">from_scratch=</span><span class="cn">TRUE</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co">#&gt; Proportions of Observations within 95% Credible Interval: 0.9721633</span></span></code></pre></div>
<p>Now we are going to finally look at the posterior distribution of our
regression parameters, but because there are so many we will focus on
just those that have a 95% credible interval not including zero (i.e.,
those that the model is fairly certain are non-zero). We are also going
to ignore the intercept term and just look at parameters associated with
age and disease status.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>posterior_summary <span class="ot">&lt;-</span> <span class="fu">summary</span>(posterior, <span class="at">pars=</span><span class="st">&quot;Lambda&quot;</span>)<span class="sc">$</span>Lambda</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>focus <span class="ot">&lt;-</span> posterior_summary[<span class="fu">sign</span>(posterior_summary<span class="sc">$</span>p2<span class="fl">.5</span>) <span class="sc">==</span> <span class="fu">sign</span>(posterior_summary<span class="sc">$</span>p97<span class="fl">.5</span>),]</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>focus <span class="ot">&lt;-</span> <span class="fu">unique</span>(focus<span class="sc">$</span>coord)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="fu">plot</span>(posterior, <span class="at">par=</span><span class="st">&quot;Lambda&quot;</span>, <span class="at">focus.coord =</span> focus, <span class="at">focus.cov =</span> <span class="fu">rownames</span>(X)[<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co">#&gt; Scale for &#39;colour&#39; is already present. Adding another scale for &#39;colour&#39;, which will</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="co">#&gt; replace the existing scale.</span></span></code></pre></div>
<p><img src="https://github.com/jsilve24/fido/raw/master/vignettes/fido-lambda-res.png" width="50%" style="display: block; margin: auto;" /></p>
<p>The first, and most obvious ting to notice is that the covariate
<code>age</code> has pretty much no effect at all, whatever effect it
may have is incredibly weak. So we are going to remove age from the plot
and just look at those coordinates with non-zero effect for diagnosis
CD</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>posterior_summary <span class="ot">&lt;-</span> <span class="fu">filter</span>(posterior_summary, covariate<span class="sc">==</span><span class="st">&quot;diagnosisCD&quot;</span>) </span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>focus <span class="ot">&lt;-</span> posterior_summary[<span class="fu">sign</span>(posterior_summary<span class="sc">$</span>p2<span class="fl">.5</span>) <span class="sc">==</span> <span class="fu">sign</span>(posterior_summary<span class="sc">$</span>p97<span class="fl">.5</span>),]</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>focus <span class="ot">&lt;-</span> <span class="fu">unique</span>(focus<span class="sc">$</span>coord)</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="fu">tax_table</span>(dat)[<span class="fu">taxa_names</span>(dat)[<span class="fu">which</span>(<span class="fu">names_coords</span>(posterior) <span class="sc">%in%</span> focus)]]</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#&gt; Taxonomy Table:     [12 taxa by 7 taxonomic ranks]:</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt;         Kingdom    Phylum           Class                   Order              </span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co">#&gt; 74305   &quot;Bacteria&quot; &quot;Proteobacteria&quot; &quot;Epsilonproteobacteria&quot; &quot;Campylobacterales&quot;</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co">#&gt; 4449236 &quot;Bacteria&quot; &quot;Proteobacteria&quot; &quot;Betaproteobacteria&quot;    &quot;Burkholderiales&quot;  </span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co">#&gt; 1105919 &quot;Bacteria&quot; &quot;Proteobacteria&quot; &quot;Betaproteobacteria&quot;    &quot;Burkholderiales&quot;  </span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co">#&gt; 4477696 &quot;Bacteria&quot; &quot;Proteobacteria&quot; &quot;Gammaproteobacteria&quot;   &quot;Pasteurellales&quot;   </span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co">#&gt; 4448331 &quot;Bacteria&quot; &quot;Proteobacteria&quot; &quot;Gammaproteobacteria&quot;   &quot;Enterobacteriales&quot;</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co">#&gt; 4154872 &quot;Bacteria&quot; &quot;Bacteroidetes&quot;  &quot;Flavobacteriia&quot;        &quot;Flavobacteriales&quot; </span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="co">#&gt; 4452538 &quot;Bacteria&quot; &quot;Fusobacteria&quot;   &quot;Fusobacteriia&quot;         &quot;Fusobacteriales&quot;  </span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="co">#&gt; 341322  &quot;Bacteria&quot; &quot;Firmicutes&quot;     &quot;Bacilli&quot;               &quot;Turicibacterales&quot; </span></span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="co">#&gt; 1015143 &quot;Bacteria&quot; &quot;Firmicutes&quot;     &quot;Bacilli&quot;               &quot;Gemellales&quot;       </span></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a><span class="co">#&gt; 176318  &quot;Bacteria&quot; &quot;Firmicutes&quot;     &quot;Clostridia&quot;            &quot;Clostridiales&quot;    </span></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a><span class="co">#&gt; 1788466 &quot;Bacteria&quot; &quot;Firmicutes&quot;     &quot;Clostridia&quot;            &quot;Clostridiales&quot;    </span></span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a><span class="co">#&gt; 1896700 &quot;Bacteria&quot; &quot;Firmicutes&quot;     &quot;Clostridia&quot;            &quot;Clostridiales&quot;    </span></span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a><span class="co">#&gt;         Family                  Genus Species</span></span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a><span class="co">#&gt; 74305   &quot;Helicobacteraceae&quot;     NA    NA     </span></span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a><span class="co">#&gt; 4449236 &quot;Alcaligenaceae&quot;        NA    NA     </span></span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a><span class="co">#&gt; 1105919 &quot;Oxalobacteraceae&quot;      NA    NA     </span></span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a><span class="co">#&gt; 4477696 &quot;Pasteurellaceae&quot;       NA    NA     </span></span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a><span class="co">#&gt; 4448331 &quot;Enterobacteriaceae&quot;    NA    NA     </span></span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a><span class="co">#&gt; 4154872 &quot;[Weeksellaceae]&quot;       NA    NA     </span></span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a><span class="co">#&gt; 4452538 &quot;Fusobacteriaceae&quot;      NA    NA     </span></span>
<span id="cb15-28"><a href="#cb15-28" tabindex="-1"></a><span class="co">#&gt; 341322  &quot;Turicibacteraceae&quot;     NA    NA     </span></span>
<span id="cb15-29"><a href="#cb15-29" tabindex="-1"></a><span class="co">#&gt; 1015143 &quot;Gemellaceae&quot;           NA    NA     </span></span>
<span id="cb15-30"><a href="#cb15-30" tabindex="-1"></a><span class="co">#&gt; 176318  &quot;Christensenellaceae&quot;   NA    NA     </span></span>
<span id="cb15-31"><a href="#cb15-31" tabindex="-1"></a><span class="co">#&gt; 1788466 &quot;Lachnospiraceae&quot;       NA    NA     </span></span>
<span id="cb15-32"><a href="#cb15-32" tabindex="-1"></a><span class="co">#&gt; 1896700 &quot;Peptostreptococcaceae&quot; NA    NA</span></span>
<span id="cb15-33"><a href="#cb15-33" tabindex="-1"></a><span class="fu">plot</span>(posterior, <span class="at">par=</span><span class="st">&quot;Lambda&quot;</span>, <span class="at">focus.coord =</span> focus, <span class="at">focus.cov =</span> <span class="fu">rownames</span>(X)[<span class="dv">2</span>])</span>
<span id="cb15-34"><a href="#cb15-34" tabindex="-1"></a><span class="co">#&gt; Scale for &#39;colour&#39; is already present. Adding another scale for &#39;colour&#39;, which will</span></span>
<span id="cb15-35"><a href="#cb15-35" tabindex="-1"></a><span class="co">#&gt; replace the existing scale.</span></span></code></pre></div>
<p><img src="https://github.com/jsilve24/fido/raw/master/vignettes/fido-lambda-cd.png" width="50%" style="display: block; margin: auto;" /></p>
</div>
<div id="more-technical-details" class="section level1">
<h1>More Technical Details</h1>
<div id="a-few-notes-on-model-inference-and-parameter-collapsing"
class="section level2">
<h2>A few notes on model inference and parameter collapsing</h2>
<p>Along with some algorithmic speed-ups enabled by the C++ Eigen
library <em>fido</em> uses conjugate priors for the regression component
of the model allowing the last three lines of the model to be collapsed
into 1 line. After this the last three lines of the model can be
re-expanded using fully conjugate sampling schemes that do not require
optimization or MCMC (only matrix operations).</p>
<p><strong>Here are the details:</strong> The collapsed model is given
by <span class="math display">\[
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j, n_j\right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\eta_j &amp;\sim T_{(D-1)\times N}(\upsilon, \Theta X, \Xi, I_N + X^T
\Gamma X)
\end{align}
\]</span> where <span class="math inline">\(A=(I_N + X^T \Gamma,
X)^{-1}\)</span> and <span class="math inline">\(T_{(D-1)\times
N}\)</span> refers to the Matrix T-distribution the <span
class="math inline">\((D-1)\times N\)</span> matrix <span
class="math inline">\(\eta\)</span> with log density given by <span
class="math display">\[\log T_{(D-1)\times N}(\eta | \upsilon, \Theta X,
\Xi, A) \propto -\frac{\upsilon+N-D-2}{2}\log |
I_{D-1}+\Xi^{-1}(\eta-\Theta X)A(\eta-\Theta X)^T |.\]</span> Rather
than using MCMC to sample <span class="math inline">\(\eta\)</span> fido
uses MAP estimation. Additionally, <em>fido</em> allows quantification
of uncertainty in MAP estimates using a Laplace approximation. We found
that in practice this MAP based Laplace approximation produced
comparable results to a full MCMC sampler but with tremendous
improvements in compute time.</p>
<p>Once samples of <span class="math inline">\(\eta\)</span> are
produced using the Laplace approximation closed form solutions for the
conditional density of <span class="math inline">\(\Lambda\)</span> and
<span class="math inline">\(\Sigma\)</span> given <span
class="math inline">\(\eta\)</span> are used to “uncollapse” the
collapsed model and produce posterior samples from the target model.
This uncollapsing is fast and given by the following matrix
equations:</p>
<p><span class="math display">\[
\begin{align}
\upsilon_N &amp;= \upsilon+N \\
\Gamma_N &amp;= (XX^T+\Gamma^{-1})^{-1} \\
\Theta_N &amp;= (\eta X^T+\Theta\Gamma^{-1})\Gamma_N \\
\Xi_N &amp;= \Xi + (\eta - \Theta_N X)(\eta - \Theta_N X)^T + (\Theta_N
- \Theta)\Gamma(\Theta_N- \Theta)^T \\
p(\Sigma | \eta, X) &amp;= W^{-1}(\Xi_N, \upsilon_N)\\
p(\Lambda | \Sigma, \eta, X) &amp;= MN_{(D-1)\times Q}(\Lambda_N,
\Sigma, \Gamma_N).
\end{align}
\]</span> If Laplace approximation is too slow, unstable (see below) or
simply not needed, the default behavior of <em>pibble</em> is to preform
the above matrix calculations and produce a single point estimate of
<span class="math inline">\(\Sigma\)</span> and <span
class="math inline">\(\Lambda\)</span> based on the posterior means of
<span class="math inline">\(p(\Sigma | \eta, X)\)</span> and <span
class="math inline">\((\Lambda | \Sigma, \eta, X)\)</span>.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-aitchison1986" class="csl-entry">
Aitchison, J. 1986. <em>The Statistical Analysis of Compositional
Data</em>. Book. Monographs on Statistics and Applied Probability.
London ; New York: Chapman; Hall.
</div>
<div id="ref-gevers2014" class="csl-entry">
Gevers, Dirk, Subra Kugathasan, Lee A Denson, Yoshiki Vázquez-Baeza,
Will Van Treuren, Boyu Ren, Emma Schwager, et al. 2014. <span>“The
Treatment-Naive Microbiome in New-Onset Crohn’s Disease.”</span>
<em>Cell Host &amp; Microbe</em> 15 (3): 382–92.
</div>
<div id="ref-silverman2018" class="csl-entry">
Silverman, Justin D, Heather Durand, Rachael J Bloom, Sayan Mukherjee,
and Lawrence A David. 2018. <span>“Dynamic Linear Models Guide Design
and Analysis of Microbiota Studies Within Artificial Human Guts.”</span>
<em>bioRxiv</em>. <a
href="https://doi.org/10.1101/306597">https://doi.org/10.1101/306597</a>.
</div>
<div id="ref-silverman2019" class="csl-entry">
Silverman, Justin D., Kimberly Roche, Zachary C. Holmes, Lawrence A.
David, and Sayan Mukherjee. 2019. <span>“<span class="nocase">Bayesian
Multinomial Logistic Normal Models through Marginally Latent Matrix-T
Processes</span>.”</span> <em>arXiv e-Prints</em>, March,
arXiv:1903.11695. <a
href="https://arxiv.org/abs/1903.11695">https://arxiv.org/abs/1903.11695</a>.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>These are very large objects with many posterior
samples, so it can take a little time to compute. Faster implementations
of summary may be included as a future update if need arises<a
href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>That said, due to the need to transform back and forth
from the default coordinate system, it is fastest to call refit on
<code>pibblefit</code> objects in the default coordinate system
bypassing these transforms.<a href="#fnref2"
class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>This can also be used to plot samples of the prior
predictive distribution if Y is null in the object as in our
<code>priors</code> object<a href="#fnref3"
class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Justin Silverman, Michelle Nixon.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>
