<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Joint Modeling  (e.g., Multiomics) with fido::Orthus • fido</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Joint Modeling  (e.g., Multiomics) with fido::Orthus">
<meta property="og:description" content="fido">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fido</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://github.com/jsilve24/fido/wiki/Installation-Details">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction-to-fido.html">Intro to fido through fido::pibble</a>
    </li>
    <li>
      <a href="../articles/non-linear-models.html">Non-Linear Modeling with fido::basset</a>
    </li>
    <li>
      <a href="../articles/orthus.html">Joint Modeling  (e.g., Multiomics) with fido::Orthus</a>
    </li>
    <li>
      <a href="../articles/picking_priors.html">Tips on Specifying Priors</a>
    </li>
    <li>
      <a href="../articles/mitigating-pcrbias.html">Mitigating PCR Bias</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://github.com/jsilve24/RcppCoDA">RcppCoDA</a>
    </li>
    <li>
      <a href="https://jsilve24.github.io/driver/">driver</a>
    </li>
    <li>
      <a href="https://bioconductor.org/packages/release/bioc/html/philr.html">philr</a>
    </li>
    <li>
      <a href="https://cran.r-project.org/package=RcppHungarian">RcppHungarian</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/inschool4life">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/jsilve24/fido">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="orthus_files/header-attrs-2.3/header-attrs.js"></script><script src="orthus_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Joint Modeling (e.g., Multiomics) with fido::Orthus</h1>
                        <h4 class="author">Justin Silverman</h4>
            
      
      
      <div class="hidden name"><code>orthus.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">fido</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">phyloseq</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">driver</span>)

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">48482</span>)</pre></body></html></div>
<p>If you have not already done so, I would read through the <em>pibble</em> vignette before this one.</p>
<div id="fidoorthus-the-model" class="section level1">
<h1 class="hasAnchor">
<a href="#fidoorthus-the-model" class="anchor"></a>fido::orthus, the model</h1>
<p>fido can be used for jointly modeling multivariate count data and multivariate Gaussian data. For example, this would be a reasonable model to jointly model 16S microbiome data and metabolomics data jointly. Because of the “two-headed” nature of this model, e.g., two observed data-sets, I named this model <em>orthus</em> (a two-headed dog and brother of Cerberus in Greek Mythology)[<a href="https://en.Wikipedia.org/wiki/Orthrus" class="uri">https://en.Wikipedia.org/wiki/Orthrus</a>]. The <em>orthus</em> model can be written as</p>
<p><span class="math display">\[
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j \right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\begin{bmatrix}\eta_j \\ Z_j \end{bmatrix} &amp;\sim N(\Lambda X, \Sigma) \\
\Lambda &amp;\sim  N(\Theta, \Sigma, \Gamma) \\
\Sigma &amp;\sim W^{-1}(\Xi, \upsilon) 
\end{align}
\]</span></p>
<p>Note this looks nearly identical to the <em>pibble</em> model but we have appended the second (Gaussian) dataset (<span class="math inline">\(Z\)</span>) onto <span class="math inline">\(\eta\)</span>. In doing this, the definition of <span class="math inline">\(\Lambda\)</span> changes (it is now larger with the bottom rows dictating how the covariates <span class="math inline">\(X\)</span> influence the second dataset). Similarly, <span class="math inline">\(\Sigma\)</span> now is much larger and can be though of as <span class="math display">\[
\Sigma = \begin{bmatrix} \Sigma_{(\eta, \eta)} &amp; \Sigma_{(\eta, Z)} \\
                          \Sigma_{(Z, \eta)} &amp; \Sigma_{(Z, Z)}\end{bmatrix}
\]</span> where <span class="math inline">\(\Sigma_{(\eta, \eta)}\)</span> describes the covariance between log-ratios (e.g., the covariance among the multinomial categories in log-ratio space), <span class="math inline">\(\Sigma_{(Z, Z)}\)</span> describes the covariance between the dimensions of <span class="math inline">\(Z\)</span> (e.g., between metabolites if Z is metabolomics data), and <span class="math inline">\(\Sigma_{(\eta, Z)} = \Sigma_{(Z, \eta)}^T\)</span> represents the covariance between log-ratios and dimensions of <span class="math inline">\(Z\)</span> (e.g., between microbial taxa and metabolites). Similar to <span class="math inline">\(\Sigma\)</span> and <span class="math inline">\(\Lambda\)</span>, the parameters <span class="math inline">\(\Xi\)</span> and <span class="math inline">\(\Theta\)</span> undergo a similar expansion to accommodate the second dataset.</p>
</div>
<div id="joint-modeling-of-microbial-16s-data-and-metabolomics" class="section level1">
<h1 class="hasAnchor">
<a href="#joint-modeling-of-microbial-16s-data-and-metabolomics" class="anchor"></a>Joint modeling of Microbial 16S data and Metabolomics</h1>
<p>To demonstrate <em>orthus</em> I will perform a toy analysis on data from <span class="citation">Kashyap et al. (2013)</span> and made available by <span class="citation">Callahan et al. (2016)</span> as part of their recently published microbiome data analysis workflow <span class="citation">(Callahan et al. 2016)</span>. I follow the data preprocessing of <span class="citation">Callahan et al. (2016)</span> we just don’t drop taxa but instead amalgamate those that don’t pass filtering to a category called “other”. I do this to maintain the proper variance in the multinomial model.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">metab_path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/Kashyap2013"</span>, <span class="st">"metabolites.csv"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"fido"</span>)
<span class="no">microbe_path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/Kashyap2013"</span>, <span class="st">"microbe.rda"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"fido"</span>)
<span class="no">metab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span>(<span class="no">metab_path</span>, <span class="kw">row.names</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="no">metab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">metab</span>)
<span class="no">microbe</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html">get</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="no">microbe_path</span>))

<span class="co">## Preprocessing ##</span>
<span class="co"># Metabolite Preprocessing</span>
<span class="no">keep_ix</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">metab</span> <span class="kw">==</span> <span class="fl">0</span>) <span class="kw">&lt;=</span> <span class="fl">3</span>
<span class="no">metab</span> <span class="kw">&lt;-</span> <span class="no">metab</span>[<span class="no">keep_ix</span>, ]

<span class="co"># 16S Preprocesing - plus some weirdness to rename amalgamated category to "other"</span>
<span class="no">keep_ix</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html">taxa_sums</a></span>(<span class="no">microbe</span>) <span class="kw">&gt;</span> <span class="fl">4</span>
<span class="no">keep_ix</span> <span class="kw">&lt;-</span> <span class="no">keep_ix</span> <span class="kw">&amp;</span> (<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span>(<span class="no">microbe</span>)<span class="kw">&gt;</span><span class="fl">2</span>)<span class="kw">&gt;</span><span class="fl">3</span>)
<span class="no">microbe</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/merge_taxa-methods.html">merge_taxa</a></span>(<span class="no">microbe</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span>(<span class="no">microbe</span>)[!<span class="no">keep_ix</span>])
<span class="no">nms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span>(<span class="no">microbe</span>)
<span class="no">rnm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span>(<span class="no">microbe</span>)<span class="kw">==</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span>(<span class="no">microbe</span>)[!<span class="no">keep_ix</span>][<span class="fl">1</span>])
<span class="no">nms</span>[<span class="no">rnm</span>] <span class="kw">&lt;-</span> <span class="st">"other"</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span>(<span class="no">microbe</span>) <span class="kw">&lt;-</span> <span class="no">nms</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span>(<span class="no">nms</span>, <span class="no">rnm</span>)

<span class="co"># bit of preprocessing </span>
<span class="no">metab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="fl">1</span> + <span class="no">metab</span>)</pre></body></html></div>
<p>Now I am going to do just a bit of processing to get data into a format for orthus. Note I have no extra metadata so we are just going to use an intercept in our model at this time.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">Y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span>(<span class="no">microbe</span>, <span class="kw">taxa_are_rows</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">Z</span> <span class="kw">&lt;-</span> <span class="no">metab</span> <span class="co">#(metabolites are rows)</span>
<span class="no">X</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">1</span>, <span class="fl">1</span>, <span class="kw pkg">phyloseq</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html">nsamples</a></span>(<span class="no">microbe</span>))

<span class="co"># save dims for easy reference</span>
<span class="no">N</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">Y</span>)
<span class="no">P</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">Z</span>)
<span class="no">Q</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">X</span>)
<span class="no">D</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">Y</span>)</pre></body></html></div>
<p>Now I am going to set up the priors. My priors are going to be similar to that of <em>pibble</em> but now we need to think about a prior for the covariance among the metabolites and between the metabolites and the log-ratios of the taxa. Remember, that priors must be defined in the <span class="math inline">\(ALR_D\)</span> (e.g., ALR with the reference being the D-th taxa; this may be changed in the future to make specifying priors more user friendly).</p>
<p>I am going to form our prior for <span class="math inline">\(\Sigma\)</span> by specifying <span class="math inline">\(\upsilon\)</span> and <span class="math inline">\(\Xi\)</span>. I will specify that I have weak prior belief that the taxa are independent in terms of their log absolute abundance. We can translate this statement about covariance of log absolute abundance into a statement about log-ratio covariance by pre- and post-multiplying by the <span class="math inline">\(ALR_D\)</span> contrast matrix (which I refer to as <span class="math inline">\(GG\)</span> below). Additionally, I believe that there is likely no substantial covariance between the taxa and the metabolites and I assume the metabolites are likely independent.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">upsilon</span> <span class="kw">&lt;-</span> (<span class="no">D</span>-<span class="fl">1</span>+<span class="no">P</span>)+<span class="fl">10</span> <span class="co"># weak-ish prior on covariance over joint taxa and metabolites</span>
<span class="no">Xi</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">D</span>-<span class="fl">1</span>+<span class="no">P</span>)
<span class="no">GG</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">D</span>-<span class="fl">1</span>), -<span class="fl">1</span>)
<span class="no">Xi</span>[<span class="fl">1</span>:(<span class="no">D</span>-<span class="fl">1</span>), <span class="fl">1</span>:(<span class="no">D</span>-<span class="fl">1</span>)] <span class="kw">&lt;-</span> <span class="no">GG</span><span class="kw">%*%</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">D</span>) <span class="kw">%*%</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/transpose-methods.html">t</a></span>(<span class="no">GG</span>)
<span class="no">Xi</span> <span class="kw">&lt;-</span> <span class="no">Xi</span> * (<span class="no">upsilon</span>-<span class="no">D</span>-<span class="no">P</span>) <span class="co"># this scales Xi to have the proper mean we wanted</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/image.html">image</a></span>(<span class="no">Xi</span>)</pre></body></html></div>
<p><img src="orthus_files/figure-html/unnamed-chunk-5-1.png" width="480"></p>
<p>Note the structure of this prior, everything is independent but there is a moderate positive covariance between the log-ratios based on their shared definition in terms of the <span class="math inline">\(D\)</span>-th taxa.</p>
<p>The other parts of the prior are less interesting. We are going to state that our mean for <span class="math inline">\(\Lambda\)</span> is centered about <span class="math inline">\(\mathbf{0}\)</span> and that the signal-to-noise ratio in the data is approximately 1 (this later part is specified by <span class="math inline">\(\Gamma=I\)</span>).</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">Gamma</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">Q</span>)
<span class="no">Theta</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">0</span>, <span class="no">D</span>-<span class="fl">1</span>+<span class="no">P</span>, <span class="fl">1</span>)</pre></body></html></div>
<p>Finally I fit the model.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/orthus_fit.html">orthus</a></span>(<span class="no">Y</span>, <span class="no">Z</span>, <span class="no">X</span>, <span class="kw">Theta</span><span class="kw">=</span><span class="no">Theta</span>, <span class="kw">Gamma</span><span class="kw">=</span><span class="no">Gamma</span>, <span class="kw">Xi</span><span class="kw">=</span><span class="no">Xi</span>, <span class="kw">upsilon</span><span class="kw">=</span><span class="no">upsilon</span>, <span class="kw">n_samples</span><span class="kw">=</span><span class="fl">1000</span>)</pre></body></html></div>
<p>Next we are going to transform the log-ratios from <span class="math inline">\(ALR_D\)</span> to the <span class="math inline">\(CLR\)</span>. I have written all the transformation functions, <em>e.g.</em>, <code>to_clr</code> etc… to work on <code>orthusfit</code> objects in a similar manner to how they work on <code>pibblefit</code> objects. For <code>orthusfit</code> objects they only transform the log-ratio components of parameters leaving the other parts of inferred model parameters (<em>i.e.</em>, the parts associated with the metabolites) untouched.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fido_transforms.html">to_clr</a></span>(<span class="no">fit</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">fit</span>)
<span class="co">#&gt; orthusfit Object: </span>
<span class="co">#&gt;   Number of Samples:      12 </span>
<span class="co">#&gt;   Number of Categories:       114 </span>
<span class="co">#&gt;   Number of Zdimensions:  405 </span>
<span class="co">#&gt;   Number of Covariates:       1 </span>
<span class="co">#&gt;   Number of Posterior Samples:    1000 </span>
<span class="co">#&gt;   Contains Samples of Parameters:Eta  Lambda  Sigma</span>
<span class="co">#&gt;   Coordinate System:      clr</span></pre></body></html></div>
</div>
<div id="investigate-model-results" class="section level1">
<h1 class="hasAnchor">
<a href="#investigate-model-results" class="anchor"></a>Investigate Model Results</h1>
<p>There are a ton of ways to visualize the inferred model. I could make network diagrams relating taxa to taxa, taxa to metabolites and metabolites to metabolites. I could look at a low dimensional representation of joint covariance to create something very much akin to canonical correlation analysis (CCA). I could look at how well the metabolites predict the taxa and vice-versa. But for the sake of simplicity I will do something much simpler. Here I am just going to find a list of taxa metabolite covariances that the model is very confident about.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co"># First just look ath the cross-covariances fit by the model</span>
<span class="co"># (covariance between taxa in CLR coordinates and metabolites)</span>
<span class="co"># This requires that we extract the corner of Sigma. </span>
<span class="no">xcor</span> <span class="kw">&lt;-</span> <span class="no">fit</span>$<span class="no">Sigma</span>[<span class="fl">1</span>:<span class="no">D</span>, <span class="no">D</span>:(<span class="no">D</span>-<span class="fl">1</span>+<span class="no">P</span>),]

<span class="co"># Initial preprocessing to speed up computation of posterior intervals</span>
<span class="co"># As there are alot of cross-covariance terms we are going to first </span>
<span class="co"># weed down the list of things we have to look at by first pass </span>
<span class="co"># selecting only those taxa that have a large posterior mean for the covariance</span>
<span class="no">xcor.mean</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">xcor</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="no">mean</span>)
<span class="no">to.analyze</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/driver/man/gather_array.html">gather_array</a></span>(<span class="no">xcor.mean</span>, <span class="no">cov</span>, <span class="no">taxa</span>, <span class="no">metabolite</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(-<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">cov</span>)) <span class="kw">%&gt;%</span>
  <span class="no">.</span>[<span class="fl">1</span>:<span class="fl">1000</span>,] <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">tm</span> <span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">taxa</span>, <span class="st">"_"</span>, <span class="no">metabolite</span>))

<span class="co"># Subset Covariance to those we are interested in and calculate posterior </span>
<span class="co"># confidence intervals. </span>
<span class="no">xcor.summary</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/driver/man/gather_array.html">gather_array</a></span>(<span class="no">xcor</span>, <span class="no">cov</span>, <span class="no">taxa</span>, <span class="no">metabolite</span>, <span class="no">iter</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">tm</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">taxa</span>, <span class="st">"_"</span>, <span class="no">metabolite</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">tm</span> <span class="kw">%in%</span> <span class="no">to.analyze</span>$<span class="no">tm</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">taxa</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">Y</span>)[<span class="no">taxa</span>], <span class="kw">metabolite</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">Z</span>)[<span class="no">metabolite</span>]) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">taxa</span>, <span class="no">metabolite</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/driver/man/summarise_posterior.html">summarise_posterior</a></span>(<span class="no">cov</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="no">mean</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">taxa</span> <span class="kw">!=</span> <span class="st">'other'</span>) <span class="co"># we don't care about these</span>

<span class="co"># Select those covariances where the model has high certainty (95%) that</span>
<span class="co"># the true covariance is not zero. </span>
<span class="no">xcor.summary</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span>(<span class="no">p2.5</span>)<span class="kw">==</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span>(<span class="no">p97.5</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">mean</span>) <span class="kw">&gt;</span> <span class="fl">2</span>)
<span class="co">#&gt; # A tibble: 214 x 8</span>
<span class="co">#&gt; # Groups:   taxa [17]</span>
<span class="co">#&gt;    taxa  metabolite   p2.5   p25   p50  mean   p75  p97.5</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 722   206.0445922 -6.54 -3.92 -3.09 -3.30 -2.39 -1.45 </span>
<span class="co">#&gt;  2 7816  206.0445922 -5.38 -3.19 -2.42 -2.66 -1.85 -1.06 </span>
<span class="co">#&gt;  3 722   290.9298419 -4.89 -3.09 -2.45 -2.60 -1.88 -1.16 </span>
<span class="co">#&gt;  4 18182 380.1846197 -4.87 -3.11 -2.36 -2.52 -1.83 -1.03 </span>
<span class="co">#&gt;  5 722   181.4504354 -4.77 -3.03 -2.36 -2.51 -1.83 -1.18 </span>
<span class="co">#&gt;  6 722   177.0565368 -4.90 -2.98 -2.34 -2.49 -1.81 -1.10 </span>
<span class="co">#&gt;  7 19517 380.1846197 -4.82 -3.02 -2.31 -2.47 -1.74 -0.920</span>
<span class="co">#&gt;  8 2943  380.1846197 -4.69 -2.99 -2.31 -2.46 -1.79 -0.994</span>
<span class="co">#&gt;  9 722   176.0343919 -4.80 -2.97 -2.28 -2.46 -1.80 -1.10 </span>
<span class="co">#&gt; 10 722   180.072273  -4.74 -3.00 -2.26 -2.45 -1.70 -1.00 </span>
<span class="co">#&gt; # … with 204 more rows</span></pre></body></html></div>
<p>So it looks there there are a few hundred covariances that we can be fairly confident about.</p>
</div>
<div id="qualifications-and-caution" class="section level1">
<h1 class="hasAnchor">
<a href="#qualifications-and-caution" class="anchor"></a>Qualifications and Caution</h1>
<p>Please note, I performed this analysis to demonstrate the use of <em>orthus</em> which is a model that I have been repeatedly asked for. I think its a cool model and could be quite useful in the right circumstances. But I would like to point out a few philosophical points about the analysis I performed above.</p>
<p>First, I performed this analysis just to demonstrate <em>orthus</em>. I really don’t know the data showcased here. What is metabolite <code>206.0445922</code>? I have no idea. For some reason this is how the metabolites in that dataset were named. For the same reason I have left the taxa indexed by sequence variant number.</p>
<p>Second (and more important), identifying relationships between taxa and metabolites (or between any two high-dimensional multivariate data-sets) is really difficult! Here we are looking at just 114 taxa and 405 but this leads to 46170 possible covariances and here we only have 12 samples! Yes <em>orthus</em> is a Bayesian model, and Yes, Bayesian models can be quite useful when there are more parameters than samples, but there is a limit of reasonability. Really, Bayesian models are great when you can perfectly capture your prior beliefs with your prior. But how often can that really be done perfectly? As such I would caution users, to use <em>orthus</em> carefully. Consider which metabolites and taxa you really care about and if you can, isolate your analyses to those.</p>
<p>Alright, that’s probably enough philosophizing for an R package Vignette. I hope you enjoy <em>orthus</em>.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-callahan2016">
<p>Callahan, Ben J, Kris Sankaran, Julia A Fukuyama, Paul J McMurdie, and Susan P Holmes. 2016. “Bioconductor Workflow for Microbiome Data Analysis: From Raw Reads to Community Analyses.” <em>F1000Research</em> 5.</p>
</div>
<div id="ref-kashyap2013">
<p>Kashyap, Purna C, Angela Marcobal, Luke K Ursell, Samuel A Smits, Erica D Sonnenburg, Elizabeth K Costello, Steven K Higginbottom, et al. 2013. “Genetically Dictated Change in Host Mucus Carbohydrate Landscape Exerts a Diet-Dependent Effect on the Gut Microbiota.” <em>Proceedings of the National Academy of Sciences</em> 110 (42): 17059–64.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Justin Silverman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
