<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to fido::Pibble • fido</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to fido::Pibble">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fido</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introduction-to-fido.html">Introduction to fido::Pibble</a></li>
    <li><a class="dropdown-item" href="../articles/non-linear-models.html">Non-linear models with fido::basset</a></li>
    <li><a class="dropdown-item" href="../articles/orthus.html">Joint Modeling  (e.g., Multiomics) with fido::Orthus</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jsilve24/fido/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to fido::Pibble</h1>
                        <h4 data-toc-skip class="author">Justin
Silverman</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jsilve24/fido/blob/master/vignettes/introduction-to-fido.Rmd" class="external-link"><code>vignettes/introduction-to-fido.Rmd</code></a></small>
      <div class="d-none name"><code>introduction-to-fido.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="an-introduction-to-fido">An introduction to <em>fido</em><a class="anchor" aria-label="anchor" href="#an-introduction-to-fido"></a>
</h2>
<p><em>fido</em> <span class="citation">(Justin D. Silverman et al.
2019)</span> is a loose acronym for <strong>“(Bayesian) Multinomial
Logistic-Normal Models”</strong>. In particular the development of
<em>fido</em> stems from the need for fast inference for time-invariant
MALLARD models<span class="citation">(Justin D. Silverman et al.
2018)</span>. <em>fido</em> is very fast! It uses closed form solutions
for model gradients and Hessians written in C++ to preform <a href="https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation" class="external-link">MAP
estimation</a> in combination with parameter uncertainty estimation
using a <a href="https://www.sumsar.net/blog/2013/11/easy-laplace-approximation/" class="external-link">Laplace
Approximation</a>. One of the main models in <em>fido</em> is the
function <em>pibble</em> which fits a Multinomial Logistic-Normal
<strong>Linear Regression</strong> model.</p>
<p><strong>So what is a <em>fido</em> model exactly?</strong> First let
me give the broad description from 10,000ft up: Basically its a model
for multinomial count data (e.g., each sample contains the counts of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>D</mi><annotation encoding="application/x-tex">D</annotation></semantics></math>
“types of things”). Importantly, unlike the more common Poisson count
models, the multinomial models a “competition to be counted” (i.e.,
cases in which counting more of one type of thing means that I have less
resources available to count other types of things).</p>
<p>This may seem vague so let me give an example. Pretend there is a
ball pit with red, green, and blue balls. Pretend that the ball pit is
very large and I don’t know the total number of balls in the ball pit,
yet I want to say something about the relative number of red, blue, and
green balls in the pit. One way I may choose to measure the ball pit is
by grabbing an armful of balls and counting the number of balls of each
color (e.g., in one armful I may collect 5 red, 3 blue, and 6 green). My
arms can only contain so many balls (in this example about 14) and so if
I were to have (randomly) gotten another green ball in my armful (making
7 total) I would likely not have been able to measure one of the red or
blue balls; hence the “competition to be counted”. It turns out that
this type of sampling occurs all the time in many situations (Wikipedia
has an example with <a href="https://en.Wikipedia.org/wiki/Multinomial_distribution#Example" class="external-link">political
polling</a>). Perhaps one of the most notable examples of this type of
count data occurs with modern high-throughput sequencing studies such as
16S rRNA studies to profile microbial communities or bulk/single-cell
RNA-seq studies to study expression profiles of cells. In all cases,
transcripts are sequenced and the number of different types of
transcripts are counted. The important part is that sequencing only
samples a small portion of the total genetic material available and
leads to similar competition to be counted.</p>
<div class="section level3">
<h3 id="the-pibble-model">The <em>pibble</em> model<a class="anchor" aria-label="anchor" href="#the-pibble-model"></a>
</h3>
<p><em>Pibble</em> is one type of <em>fido</em> model. In particular its
a <em>fido</em> model for multivariate linear regression.</p>
<p>Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
denote an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><mo>×</mo><mi>N</mi></mrow><annotation encoding="application/x-tex">D\times N</annotation></semantics></math>
matrix of counts. Let us denote the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>-th
column of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mi>j</mi></msub><annotation encoding="application/x-tex">Y_j</annotation></semantics></math>.
Thus each “sample” in the dataset is a measurement of the relative
amount of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>D</mi><annotation encoding="application/x-tex">D</annotation></semantics></math>
“types of things”. Suppose we also have have covariate information in
the form of a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Q</mi><mo>×</mo><mi>N</mi></mrow><annotation encoding="application/x-tex">Q\times N</annotation></semantics></math>
matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>.</p>
<p>The following is the pibble model including likelihood and priors:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Y</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>∼</mo><mtext mathvariant="normal">Multinomial</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>π</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>π</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mi>ϕ</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>η</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>η</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>∼</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Λ</mi><msub><mi>X</mi><mi>j</mi></msub><mo>,</mo><mi>Σ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>Λ</mi></mtd><mtd columnalign="left" style="text-align: left"><mo>∼</mo><mi>M</mi><msub><mi>N</mi><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>Q</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>Θ</mi><mo>,</mo><mi>Σ</mi><mo>,</mo><mi>Γ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>Σ</mi></mtd><mtd columnalign="left" style="text-align: left"><mo>∼</mo><msup><mi>W</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>Ξ</mi><mo>,</mo><mi>υ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j \right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\eta_j &amp;\sim N(\Lambda X_j, \Sigma) \\
\Lambda &amp;\sim  MN_{(D-1) \times Q}(\Theta, \Sigma, \Gamma) \\
\Sigma &amp;\sim W^{-1}(\Xi, \upsilon) 
\end{align}
</annotation></semantics></math> Here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>M</mi><msub><mi>N</mi><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>Q</mi></mrow></msub></mrow><annotation encoding="application/x-tex">MN_{(D-1) \times Q}</annotation></semantics></math>
denotes a <a href="https://en.wikipedia.org/wiki/Matrix_normal_distribution" class="external-link">Matrix
Normal distribution</a> for a matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
of regression coefficients of dimension
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>Q</mi></mrow><annotation encoding="application/x-tex">(D-1)\times Q</annotation></semantics></math>.
Essentially you can think of the Matrix normal as having two covariance
matrices one describing the covariation between the rows of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Σ</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>)
and another describing the covariation of the columns of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Γ</mi><annotation encoding="application/x-tex">\Gamma</annotation></semantics></math>).
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>W</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><annotation encoding="application/x-tex">W^{-1}</annotation></semantics></math>
refers to the <a href="https://en.wikipedia.org/wiki/Inverse-Wishart_distribution" class="external-link">Inverse
Wishart distribution</a> (which is a common distribution over covariance
matrices). The line
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>π</mi><mi>j</mi></msub><mo>=</mo><msup><mi>ϕ</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>η</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\pi_j = \phi^{-1}(\eta_j)</annotation></semantics></math>
represents a transformation between the parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\pi_j</annotation></semantics></math>
which exist on a simplex (e.g.,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\pi_j</annotation></semantics></math>
must sum to 1) and the transformed parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>η</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\eta_j</annotation></semantics></math>
that exist in real space. In particular we define
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>ϕ</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><annotation encoding="application/x-tex">\phi^{-1}</annotation></semantics></math>
to be the <a href="http://www.sediment.uni-goettingen.de/staff/tolosana/extra/CoDaNutshell.pdf" class="external-link">inverse
additive log ratio transform</a> (which conversely implies that
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>η</mi><mi>j</mi></msub><mo>=</mo><mi>A</mi><mi>L</mi><mi>R</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>π</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\eta_j = ALR(\pi_j)</annotation></semantics></math>)
also known as the identified softmax transform (as it is more commonly
known in the Machine Learning community). While I will say more on this
later in this tutorial, one thing to know is that I have the model
implemented using the ALR transform as it is computationally simple and
fast; the results of the model can be viewed as if any number of
transforms had been used (instead of the ALR) including the isometric
log-ratio transform, or the centered log-ratio transform.</p>
<p>Before moving on, I would like to give <strong>a more intuitive
description of <em>pibble</em></strong>. Essentially the main modeling
component of <em>pibble</em> is the third equation above
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>η</mi><mi>j</mi></msub><mo>∼</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Λ</mi><msub><mi>X</mi><mi>j</mi></msub><mo>,</mo><mi>Σ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\eta_j \sim N(\Lambda X_j, \Sigma)</annotation></semantics></math>)
which is just a multivariate linear model. That is,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
are your covariates (which can be continuous, discrete, binary, etc…),
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Σ</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>
is the covariance matrix for the regression residuals.</p>
</div>
</div>
<div class="section level2">
<h2 id="example-analysis-of-microbiome-data">Example analysis of microbiome data<a class="anchor" aria-label="anchor" href="#example-analysis-of-microbiome-data"></a>
</h2>
<p>This analysis is the same as that presented in the <em>fido</em>
manuscript <span class="citation">(Justin D. Silverman et al.
2019)</span>. I will reanalyze a previously published study comparing
microbial composition in the terminal ileum of subjects with Crohn’s
Disease (CD) to healthy controls <span class="citation">(Gevers et al.
2014)</span>. To do this I will fit a pibble model using CD status,
inflammation status and age as covariates (plus a constant intercept
term).</p>
<p>For convienece, we have added a copy of the data set to
<em>fido</em>. The data was obtained from the <em>MicrobeDS</em>
repository on <a href="https://github.com/twbattaglia/MicrobeDS" class="external-link">GitHub</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jsilve24.github.io/fido/">fido</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">899</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">RISK_CCFA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># making into a phyloseq object</span></span>
<span><span class="va">CCFA_phylo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq.html" class="external-link">phyloseq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html" class="external-link">otu_table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">RISK_CCFA_otu</span><span class="op">)</span>, taxa_are_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">RISK_CCFA_sam</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">RISK_CCFA_tax</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># drop low abundant taxa and samples</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">CCFA_phylo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">disease_stat</span><span class="op">!=</span><span class="st">"missing"</span>, </span>
<span>                 <span class="va">immunosup</span><span class="op">!=</span><span class="st">"missing"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">diagnosis</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"no"</span>, <span class="st">"CD"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">steroids</span><span class="op">==</span><span class="st">"false"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">antibiotics</span><span class="op">==</span><span class="st">"false"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">biologics</span><span class="op">==</span><span class="st">"false"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">biopsy_location</span><span class="op">==</span><span class="st">"Terminal ileum"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_glom.html" class="external-link">tax_glom</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/prune_samples-methods.html" class="external-link">prune_samples</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_sums.html" class="external-link">sample_sums</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5000</span>,<span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/filter_taxa.html" class="external-link">filter_taxa</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.10</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Create Design Matrix and OTU Table</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>,<span class="st">"matrix"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">age</span><span class="op">)</span><span class="op">)</span>,</span>
<span>         diagnosis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html" class="external-link">relevel</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">diagnosis</span>, ordered <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, ref<span class="op">=</span><span class="st">"no"</span><span class="op">)</span>, </span>
<span>         disease_stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html" class="external-link">relevel</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">disease_stat</span>, ordered <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, ref<span class="op">=</span><span class="st">"non-inflamed"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">diagnosis</span> <span class="op">+</span> <span class="va">disease_stat</span><span class="op">+</span><span class="va">age</span>, data<span class="op">=</span><span class="va">sample_dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html" class="external-link">otu_table</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Investigate X and Y look like</span></span>
<span><span class="va">X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                      1939.SKBTI.0175 1939.SKBTI047 1939.SKBTI051 1939.SKBTI063</span></span>
<span><span class="co">#&gt; (Intercept)                  1.00000       1.00000          1.00       1.00000</span></span>
<span><span class="co">#&gt; diagnosisCD                  1.00000       1.00000          1.00       1.00000</span></span>
<span><span class="co">#&gt; disease_statinflamed         0.00000       1.00000          1.00       1.00000</span></span>
<span><span class="co">#&gt; age                         15.16667      14.33333         15.75      13.58333</span></span>
<span><span class="co">#&gt;                      1939.SKBTI072</span></span>
<span><span class="co">#&gt; (Intercept)                   1.00</span></span>
<span><span class="co">#&gt; diagnosisCD                   1.00</span></span>
<span><span class="co">#&gt; disease_statinflamed          1.00</span></span>
<span><span class="co">#&gt; age                          15.75</span></span>
<span><span class="va">Y</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; OTU Table:          [5 taxa and 5 samples]</span></span>
<span><span class="co">#&gt;                      taxa are rows</span></span>
<span><span class="co">#&gt;         1939.SKBTI.0175 1939.SKBTI047 1939.SKBTI051 1939.SKBTI063 1939.SKBTI072</span></span>
<span><span class="co">#&gt; 4442127               0             9             0            14             2</span></span>
<span><span class="co">#&gt; 74305                 1             2            35             1             0</span></span>
<span><span class="co">#&gt; 663573               36             1             0             2             1</span></span>
<span><span class="co">#&gt; 2685602              10           264           211           276            83</span></span>
<span><span class="co">#&gt; 4339819               0            37            42            70            22</span></span></code></pre></div>
<p>Next specify priors. We are going to start by specifying a prior on
the covariance between log-ratios
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Σ</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>.
I like to do this by thinking about a prior on the covariance between
taxa on the log-scale (i.e., between the log of their absolute
abundances not the log-ratios). I will refer to this covariance on
log-absolute abundances
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ω</mi><annotation encoding="application/x-tex">\Omega</annotation></semantics></math>.
For example, here I will build a prior that states that the mean of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ω</mi><annotation encoding="application/x-tex">\Omega</annotation></semantics></math>
is the identity matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>D</mi></msub><annotation encoding="application/x-tex">I_D</annotation></semantics></math>.
From From <span class="citation">Aitchison (1986)</span>, we know that
if we assume that the taxa have a covariance
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ω</mi><annotation encoding="application/x-tex">\Omega</annotation></semantics></math>
in terms of log-absolute abundance then their correlation in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mtext mathvariant="normal">ALR</mtext><mi>D</mi></msub><annotation encoding="application/x-tex">\text{ALR}_D</annotation></semantics></math>
is given by
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Σ</mi><mo>=</mo><mi>G</mi><mi>Ω</mi><msup><mi>G</mi><mi>T</mi></msup></mrow><annotation encoding="application/x-tex"> \Sigma = G \Omega G^T </annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>G</mi><annotation encoding="application/x-tex">G</annotation></semantics></math>
is a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><mo>−</mo><mn>1</mn><mo>×</mo><mi>D</mi></mrow><annotation encoding="application/x-tex">D-1 \times D</annotation></semantics></math>
matrix given by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>I</mi><mrow><mi>D</mi><mo>−</mo><mn>1</mn></mrow></msub><mo>;</mo><mo>−</mo><msub><mn>1</mn><mrow><mi>D</mi><mo>−</mo><mn>1</mn></mrow></msub><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">G = [I_{D-1}; -1_{D-1}]</annotation></semantics></math>
(i.e.,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>G</mi><annotation encoding="application/x-tex">G</annotation></semantics></math>
is the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mtext mathvariant="normal">ALR</mtext><mi>D</mi></msub><annotation encoding="application/x-tex">\text{ALR}_D</annotation></semantics></math>
contrast matrix). Additionally, we know that the Inverse Wishart mode is
given by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mi>Ξ</mi><mrow><mi>υ</mi><mo>+</mo><mi>D</mi></mrow></mfrac><annotation encoding="application/x-tex">\frac{\Xi}{\upsilon + D}</annotation></semantics></math>.
Finally, note that
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>υ</mi><annotation encoding="application/x-tex">\upsilon</annotation></semantics></math>
essentially controls our uncertainty in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Σ</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>
about this prior mean. Here I will take
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>υ</mi><mo>=</mo><mi>D</mi><mo>+</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">\upsilon = D+3</annotation></semantics></math>.
This then gives us
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Ξ</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>υ</mi><mo>−</mo><mi>D</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>G</mi><mi>I</mi><msup><mi>G</mi><mi>T</mi></msup></mrow><annotation encoding="application/x-tex">\Xi = (\upsilon - D) GIG^T</annotation></semantics></math>.
We scale
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ξ</mi><annotation encoding="application/x-tex">\Xi</annotation></semantics></math>
by a factor of 1/2 to make
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>T</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Ξ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>D</mi><mo>−</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">Tr(\Xi)=D-1</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">upsilon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ntaxa-methods.html" class="external-link">ntaxa</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">+</span><span class="fl">3</span> </span>
<span><span class="va">Omega</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ntaxa-methods.html" class="external-link">ntaxa</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ntaxa-methods.html" class="external-link">ntaxa</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">Xi</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">upsilon</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ntaxa-methods.html" class="external-link">ntaxa</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="va">G</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">Omega</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span></span></code></pre></div>
<p>Finally I specify my priors for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Θ</mi><annotation encoding="application/x-tex">\Theta</annotation></semantics></math>
(mean of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>)
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Γ</mi><annotation encoding="application/x-tex">\Gamma</annotation></semantics></math>
(covariance between columns of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>;
i.e., covariance between the covariates). I will center my prior for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
about zero, and assume that the covariates are independent.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Theta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ntaxa-methods.html" class="external-link">ntaxa</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Gamma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>I strongly recommend users perform prior predictive checks to make
sure their priors make sense to them. <em>fido</em> makes this easy, all
the main fitting functions (e.g., <code>pibble</code>) will
automatically sample from the prior predictive distribution if
<code>Y</code> is left as <code>NULL</code> (e.g., without data your
posterior is just your prior).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pibble_fit.html">pibble</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="va">X</span>, <span class="va">upsilon</span>, <span class="va">Theta</span>, <span class="va">Gamma</span>, <span class="va">Xi</span><span class="op">)</span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">priors</span><span class="op">)</span></span>
<span><span class="co">#&gt;  pibblefit Object (Priors Only): </span></span>
<span><span class="co">#&gt;   Number of Samples:      250 </span></span>
<span><span class="co">#&gt;   Number of Categories:       49 </span></span>
<span><span class="co">#&gt;   Number of Covariates:       4 </span></span>
<span><span class="co">#&gt;   Number of Posterior Samples:    2000 </span></span>
<span><span class="co">#&gt;   Contains Samples of Parameters:Eta  Lambda  Sigma</span></span>
<span><span class="co">#&gt;   Coordinate System:      alr, reference category: 49</span></span></code></pre></div>
<p>The main fitting functions in the <em>fido</em> package output
special fit objects (e.g., <code>pibble</code> outputs an object of
class <code>pibblefit</code>). These fit objects are just lists with
some extra metadata that allows special method dispatch. For example, if
you call print on a <code>pibblefit</code> object you will get a nice
summary of what is in the object.</p>
<p><em>Note:</em> Currently, the function <code>pibble</code> takes
expects inputs and outputs in the “default” coordinate system; this is
simply the ALR coordinate system where the last category (49 above) is
taken as reference (this will be generalized in future versions). More
specifically for a vector
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
representing the proportions of categories
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>D</mi><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">\{1, \dots, D\}</annotation></semantics></math>
we can write
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>x</mi><mo>*</mo></msup><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><mfrac><msub><mi>x</mi><mn>1</mn></msub><msub><mi>x</mi><mi>D</mi></msub></mfrac><mo>,</mo><mi>…</mi><mo>,</mo><mo>log</mo><mfrac><msub><mi>x</mi><mrow><mi>D</mi><mo>−</mo><mn>1</mn></mrow></msub><msub><mi>x</mi><mi>D</mi></msub></mfrac><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">x^* = \left( \log \frac{x_1}{x_D}, \dots, \log \frac{x_{D-1}}{x_D}\right).</annotation></semantics></math>
As mentioned above however, I have designed <em>fido</em> to work with
many different coordinate systems including the ALR (with respect to any
category), CLR, ILR, or proportions. To help transform things between
these coordinate systems I have written a series of transformation
functions that transform any <code>pibblefit</code> object into a
desired coordinate system. Importantly, <code>pibblefit</code> objects
keep track of what coordinate system they are currently in so as a user
you only need to specify the coordinate system that you want to change
into. Keep in mind that covariance matrices cannot be represented in
proportions and so visualizations or summaries based on covariance
matrices will be suppressed when <code>pibblefit</code> objects are in
the proportions coordinate system. As an example, lets look at viewing a
summary of the prior for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
with respect to the CLR coordinate system<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;These are very large objects with many posterior
samples, so it can take a little time to compute. Faster implementations
of summary may be included as a future update if need arises&lt;/p&gt;"><sup>1</sup></a>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fido_transforms.html">to_clr</a></span><span class="op">(</span><span class="va">priors</span><span class="op">)</span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">priors</span>, pars<span class="op">=</span><span class="st">"Lambda"</span>, gather_prob<span class="op">=</span><span class="cn">TRUE</span>, as_factor<span class="op">=</span><span class="cn">TRUE</span>, use_names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>  </span>
<span><span class="co">#&gt; $Lambda</span></span>
<span><span class="co">#&gt; # A tibble: 784 × 9</span></span>
<span><span class="co">#&gt;    Parameter coord covariate         val .lower .upper .width .point .interval</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;       &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    </span></span>
<span><span class="co">#&gt;  1 Lambda        1         1  0.0509     -0.528  0.596    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  2 Lambda        1         2  0.00199    -0.567  0.575    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  3 Lambda        1         3 -0.0373     -0.620  0.532    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  4 Lambda        1         4 -0.00205    -0.554  0.549    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  5 Lambda        2         1  0.00991    -0.564  0.535    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  6 Lambda        2         2 -0.00000782 -0.572  0.580    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  7 Lambda        2         3 -0.0247     -0.570  0.518    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  8 Lambda        2         4  0.0165     -0.536  0.589    0.5 mean   qi       </span></span>
<span><span class="co">#&gt;  9 Lambda        3         1  0.0138     -0.584  0.620    0.5 mean   qi       </span></span>
<span><span class="co">#&gt; 10 Lambda        3         2 -0.00502    -0.542  0.581    0.5 mean   qi       </span></span>
<span><span class="co">#&gt; # ℹ 774 more rows</span></span></code></pre></div>
<p>By default the <code>summary</code> function returns a list (with
possible elements <code>Lambda</code>, <code>Sigma</code>, and
<code>Eta</code>) summarizing each posterior parameter based on
quantiles and mean (e.g., p2.5 is the 0.025 percentile of the posterior
distribution). As this type of table may be hard to take in due to how
large it is, <code>pibblefit</code> objects also come with a default
plotting option for each of the parameters. Also the returned plot
objects are <code>ggplot</code> objects so normal <code>ggplot2</code>
commands work on them. Before doing that though we are going to use one
of the <code>names</code> functions for <code>pibblefit</code> objects
to provide some more specific names for the covariates (helpful when we
then plot).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/name_dims.html">names_covariates</a></span><span class="op">(</span><span class="va">priors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">priors</span>, par<span class="op">=</span><span class="st">"Lambda"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>  </span></code></pre></div>
<p><img src="pibble-priors.png"></p>
<p>This looks fairly reasonable to me. So I am going to go ahead and fit
the model with data. <code>fido</code> provides a helper method called
<code>refit</code> that we will use to avoid passing prior parameters
again.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span><span class="op">$</span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">Y</span> <span class="co"># remember pibblefit objects are just lists</span></span>
<span><span class="va">posterior</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refit.html">refit</a></span><span class="op">(</span><span class="va">priors</span>, optim_method<span class="op">=</span><span class="st">"lbfgs"</span><span class="op">)</span></span></code></pre></div>
<p>Unlike the main <em>pibble</em> function, the <code>refit</code>
method can be called on objects in any coordinate system and all
transformations to and from the default coordinate system are handled
internally<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;That said, due to the need to transform back and forth
from the default coordinate system, it is fastest to call refit on
&lt;code&gt;pibblefit&lt;/code&gt; objects in the default coordinate system
bypassing these transforms.&lt;/p&gt;"><sup>2</sup></a>. This is one nice thing about using the
<code>refit</code> method. That said, new objects added to the
<code>pibblefit</code> object need to be added in the proper coordinates
For example, if we wanted to replace our prior for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ξ</mi><annotation encoding="application/x-tex">\Xi</annotation></semantics></math>
for an object in CLR coordinates, we would had to transform our prior
for <code>Xi</code> to CLR coordinates before adding it to the
<code>priors</code> object.</p>
<p>Now I are also going to add in the taxa names to make it easier to
interpret the results.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Class"</span>, <span class="st">"Family"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">tax</span>, <span class="fl">1</span>, <span class="va">paste</span>, collapse<span class="op">=</span><span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/name_dims.html">names_categories</a></span><span class="op">(</span><span class="va">posterior</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tax</span></span></code></pre></div>
<p>Before doing anything else lets look at the posterior predictive
distribution to assess model fit. This can be accessed through the
method <code>ppc</code><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;This can also be used to plot samples of the prior
predictive distribution if Y is null in the object as in our
&lt;code&gt;priors&lt;/code&gt; object&lt;/p&gt;"><sup>3</sup></a>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ppc.html">ppc</a></span><span class="op">(</span><span class="va">posterior</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">30000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="pibble-ppc.png"></p>
<p>There are a few things to note about this plot. First, when zoomed
out like this it looks it is hard to make much of it. This is a fairly
large dataset we are analyzing and its hard to view an uncertainty
interval; in this case its plotting the median and 95% confidence
interval in grey and black and the observed counts in green.
<em>fido</em> also has a simpler function that summarizes the posterior
predictive check.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ppc_summary.html">ppc_summary</a></span><span class="op">(</span><span class="va">posterior</span><span class="op">)</span></span>
<span><span class="co">#&gt; Proportions of Observations within 95% Credible Interval: 0.9897143</span></span></code></pre></div>
<p>Here we see that the model appears to be fitting well (at least based
on the posterior predictive check) and that only about 1.5% of
observations fall outside of the 95% posterior predictive density (this
is good).</p>
<p>Some readers will look at the above <code>ppc</code> plots and think
“looks like over-fitting”. However, note that there are two ways of
using <code>ppc</code>. One is to predict the counts based on the
samples of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>η</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
(Eta; as we did above); the other is to predict “from scratch” that is
to predict starting form the posterior samples of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
(Lambda) then sampling
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>η</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
and only then sampling
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>.
This later functionality can be accessed by also passing the parameters
<code>from_scratch=TRUE</code> to the <code>ppc</code> function. Note:
these two posterior predictive checks have different meanings, one is
not better than the other.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ppc.html">ppc</a></span><span class="op">(</span><span class="va">posterior</span>, from_scratch<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">30000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="pibble-ppc-sum.png"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ppc_summary.html">ppc_summary</a></span><span class="op">(</span><span class="va">posterior</span>, from_scratch<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Proportions of Observations within 95% Credible Interval: 0.9725714</span></span></code></pre></div>
<p>Now we are going to finally look at the posterior distribution of our
regression parameters, but because there are so many we will focus on
just those that have a 95% credible interval not including zero (i.e.,
those that the model is fairly certain are non-zero). We are also going
to ignore the intercept term and just look at parameters associated with
age and disease status.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">posterior_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">posterior</span>, pars<span class="op">=</span><span class="st">"Lambda"</span><span class="op">)</span><span class="op">$</span><span class="va">Lambda</span></span>
<span><span class="va">focus</span> <span class="op">&lt;-</span> <span class="va">posterior_summary</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">posterior_summary</span><span class="op">$</span><span class="va">p2.5</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">posterior_summary</span><span class="op">$</span><span class="va">p97.5</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">focus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">focus</span><span class="op">$</span><span class="va">coord</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">posterior</span>, par<span class="op">=</span><span class="st">"Lambda"</span>, focus.coord <span class="op">=</span> <span class="va">focus</span>, focus.cov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="pibble-CD-clr.png"></p>
<p>The first, and most obvious ting to notice is that the covariate
<code>age</code> has pretty much no effect at all, whatever effect it
may have is incredibly weak. So we are going to remove age from the plot
and just look at those coordinates with non-zero effect for diagnosis
CD</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">posterior_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">posterior_summary</span>, <span class="va">covariate</span><span class="op">==</span><span class="st">"diagnosisCD"</span><span class="op">)</span> </span>
<span><span class="va">focus</span> <span class="op">&lt;-</span> <span class="va">posterior_summary</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">posterior_summary</span><span class="op">$</span><span class="va">p2.5</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">posterior_summary</span><span class="op">$</span><span class="va">p97.5</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">focus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">focus</span><span class="op">$</span><span class="va">coord</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="../reference/name_dims.html">names_coords</a></span><span class="op">(</span><span class="va">posterior</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">focus</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Taxonomy Table:     [13 taxa by 7 taxonomic ranks]:</span></span>
<span><span class="co">#&gt;         Kingdom    Phylum           Class                   Order               </span></span>
<span><span class="co">#&gt; 74305   "Bacteria" "Proteobacteria" "Epsilonproteobacteria" "Campylobacterales" </span></span>
<span><span class="co">#&gt; 4449236 "Bacteria" "Proteobacteria" "Betaproteobacteria"    "Burkholderiales"   </span></span>
<span><span class="co">#&gt; 1105919 "Bacteria" "Proteobacteria" "Betaproteobacteria"    "Burkholderiales"   </span></span>
<span><span class="co">#&gt; 4477696 "Bacteria" "Proteobacteria" "Gammaproteobacteria"   "Pasteurellales"    </span></span>
<span><span class="co">#&gt; 4448331 "Bacteria" "Proteobacteria" "Gammaproteobacteria"   "Enterobacteriales" </span></span>
<span><span class="co">#&gt; 4154872 "Bacteria" "Bacteroidetes"  "Flavobacteriia"        "Flavobacteriales"  </span></span>
<span><span class="co">#&gt; 4452538 "Bacteria" "Fusobacteria"   "Fusobacteriia"         "Fusobacteriales"   </span></span>
<span><span class="co">#&gt; 341322  "Bacteria" "Firmicutes"     "Bacilli"               "Turicibacterales"  </span></span>
<span><span class="co">#&gt; 1015143 "Bacteria" "Firmicutes"     "Bacilli"               "Gemellales"        </span></span>
<span><span class="co">#&gt; 176318  "Bacteria" "Firmicutes"     "Clostridia"            "Clostridiales"     </span></span>
<span><span class="co">#&gt; 1788466 "Bacteria" "Firmicutes"     "Clostridia"            "Clostridiales"     </span></span>
<span><span class="co">#&gt; 1896700 "Bacteria" "Firmicutes"     "Clostridia"            "Clostridiales"     </span></span>
<span><span class="co">#&gt; 191718  "Bacteria" "Firmicutes"     "Erysipelotrichi"       "Erysipelotrichales"</span></span>
<span><span class="co">#&gt;         Family                  Genus Species</span></span>
<span><span class="co">#&gt; 74305   "Helicobacteraceae"     NA    NA     </span></span>
<span><span class="co">#&gt; 4449236 "Alcaligenaceae"        NA    NA     </span></span>
<span><span class="co">#&gt; 1105919 "Oxalobacteraceae"      NA    NA     </span></span>
<span><span class="co">#&gt; 4477696 "Pasteurellaceae"       NA    NA     </span></span>
<span><span class="co">#&gt; 4448331 "Enterobacteriaceae"    NA    NA     </span></span>
<span><span class="co">#&gt; 4154872 "[Weeksellaceae]"       NA    NA     </span></span>
<span><span class="co">#&gt; 4452538 "Fusobacteriaceae"      NA    NA     </span></span>
<span><span class="co">#&gt; 341322  "Turicibacteraceae"     NA    NA     </span></span>
<span><span class="co">#&gt; 1015143 "Gemellaceae"           NA    NA     </span></span>
<span><span class="co">#&gt; 176318  "Christensenellaceae"   NA    NA     </span></span>
<span><span class="co">#&gt; 1788466 "Lachnospiraceae"       NA    NA     </span></span>
<span><span class="co">#&gt; 1896700 "Peptostreptococcaceae" NA    NA     </span></span>
<span><span class="co">#&gt; 191718  "Erysipelotrichaceae"   NA    NA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">posterior</span>, par<span class="op">=</span><span class="st">"Lambda"</span>, focus.coord <span class="op">=</span> <span class="va">focus</span>, focus.cov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="pibble-posterior-summary.png"></p>
</div>
<div class="section level2">
<h2 id="more-technical-details">More Technical Details<a class="anchor" aria-label="anchor" href="#more-technical-details"></a>
</h2>
<div class="section level3">
<h3 id="a-few-notes-on-model-inference-and-parameter-collapsing">A few notes on model inference and parameter collapsing<a class="anchor" aria-label="anchor" href="#a-few-notes-on-model-inference-and-parameter-collapsing"></a>
</h3>
<p>Along with some algorithmic speed-ups enabled by the C++ Eigen
library <em>fido</em> uses conjugate priors for the regression component
of the model allowing the last three lines of the model to be collapsed
into 1 line. After this the last three lines of the model can be
re-expanded using fully conjugate sampling schemes that do not require
optimization or MCMC (only matrix operations).</p>
<p><strong>Here are the details:</strong> The collapsed model is given
by
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Y</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>∼</mo><mtext mathvariant="normal">Multinomial</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>π</mi><mi>j</mi></msub><mo>,</mo><msub><mi>n</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>π</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mi>ϕ</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>η</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>η</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>∼</mo><msub><mi>T</mi><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>N</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>υ</mi><mo>,</mo><mi>Θ</mi><mi>X</mi><mo>,</mo><mi>Ξ</mi><mo>,</mo><msub><mi>I</mi><mi>N</mi></msub><mo>+</mo><msup><mi>X</mi><mi>T</mi></msup><mi>Γ</mi><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j, n_j\right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\eta_j &amp;\sim T_{(D-1)\times N}(\upsilon, \Theta X, \Xi, I_N + X^T \Gamma X)
\end{align}
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>I</mi><mi>N</mi></msub><mo>+</mo><msup><mi>X</mi><mi>T</mi></msup><mi>Γ</mi><mo>,</mo><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mo>−</mo><mn>1</mn></mrow></msup></mrow><annotation encoding="application/x-tex">A=(I_N + X^T \Gamma, X)^{-1}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>T</mi><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>N</mi></mrow></msub><annotation encoding="application/x-tex">T_{(D-1)\times N}</annotation></semantics></math>
refers to the Matrix T-distribution the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>N</mi></mrow><annotation encoding="application/x-tex">(D-1)\times N</annotation></semantics></math>
matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>η</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
with log density given by
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><msub><mi>T</mi><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>N</mi></mrow></msub><mo stretchy="false" form="prefix">(</mo><mi>η</mi><mrow><mo stretchy="true" form="prefix">|</mo><mi>υ</mi><mo>,</mo><mi>Θ</mi><mi>X</mi><mo>,</mo><mi>Ξ</mi><mo>,</mo><mi>A</mi><mo stretchy="false" form="postfix">)</mo><mo>∝</mo><mo>−</mo><mfrac><mrow><mi>υ</mi><mo>+</mo><mi>N</mi><mo>−</mo><mi>D</mi><mo>−</mo><mn>2</mn></mrow><mn>2</mn></mfrac><mo>log</mo><mo stretchy="true" form="postfix">|</mo></mrow><msub><mi>I</mi><mrow><mi>D</mi><mo>−</mo><mn>1</mn></mrow></msub><mo>+</mo><msup><mi>Ξ</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>η</mi><mo>−</mo><mi>Θ</mi><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>A</mi><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>η</mi><mo>−</mo><mi>Θ</mi><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>T</mi></msup><mo stretchy="false" form="prefix">|</mo><mi>.</mi></mrow><annotation encoding="application/x-tex">\log T_{(D-1)\times N}(\eta | \upsilon, \Theta X, \Xi, A) \propto -\frac{\upsilon+N-D-2}{2}\log | I_{D-1}+\Xi^{-1}(\eta-\Theta X)A(\eta-\Theta X)^T |.</annotation></semantics></math>
Rather than using MCMC to sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>η</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
fido uses MAP estimation (using a custom C++ Eigen based implementation
of the ADAM optimizer and closed form solutions for gradient and hessian
of the collapsed model)<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Which we found preformed substantially better than
L-BFGS, which we also tried.&lt;/p&gt;"><sup>4</sup></a>. Additionally, <em>fido</em> allows
quantification of uncertainty in MAP estimates using a Laplace
approximation. We found that in practice this MAP based Laplace
approximation produced comparable results to a full MCMC sampler but
with tremendous improvements in compute time.</p>
<p>Once samples of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>η</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
are produced using the Laplace approximation closed form solutions for
the conditional density of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Σ</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>
given
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>η</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
are used to “uncollapse” the collapsed model and produce posterior
samples from the target model. This uncollapsing is fast and given by
the following matrix equations:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>υ</mi><mi>N</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>υ</mi><mo>+</mo><mi>N</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Γ</mi><mi>N</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>X</mi><msup><mi>X</mi><mi>T</mi></msup><mo>+</mo><msup><mi>Γ</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mo>−</mo><mn>1</mn></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Θ</mi><mi>N</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>η</mi><msup><mi>X</mi><mi>T</mi></msup><mo>+</mo><mi>Θ</mi><msup><mi>Γ</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mo stretchy="true" form="postfix">)</mo></mrow><msub><mi>Γ</mi><mi>N</mi></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Ξ</mi><mi>N</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>Ξ</mi><mo>+</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>η</mi><mo>−</mo><msub><mi>Θ</mi><mi>N</mi></msub><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>η</mi><mo>−</mo><msub><mi>Θ</mi><mi>N</mi></msub><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>T</mi></msup><mo>+</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Θ</mi><mi>N</mi></msub><mo>−</mo><mi>Θ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>Γ</mi><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Θ</mi><mi>N</mi></msub><mo>−</mo><mi>Θ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>T</mi></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>p</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Σ</mi><mo stretchy="false" form="prefix">|</mo><mi>η</mi><mo>,</mo><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mi>W</mi><mrow><mo>−</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Ξ</mi><mi>N</mi></msub><mo>,</mo><msub><mi>υ</mi><mi>N</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>p</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Λ</mi><mo stretchy="false" form="prefix">|</mo><mi>Σ</mi><mo>,</mo><mi>η</mi><mo>,</mo><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>M</mi><msub><mi>N</mi><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>D</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>×</mo><mi>Q</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Λ</mi><mi>N</mi></msub><mo>,</mo><mi>Σ</mi><mo>,</mo><msub><mi>Γ</mi><mi>N</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
\upsilon_N &amp;= \upsilon+N \\
\Gamma_N &amp;= (XX^T+\Gamma^{-1})^{-1} \\
\Theta_N &amp;= (\eta X^T+\Theta\Gamma^{-1})\Gamma_N \\
\Xi_N &amp;= \Xi + (\eta - \Theta_N X)(\eta - \Theta_N X)^T + (\Theta_N - \Theta)\Gamma(\Theta_N- \Theta)^T \\
p(\Sigma | \eta, X) &amp;= W^{-1}(\Xi_N, \upsilon_N)\\
p(\Lambda | \Sigma, \eta, X) &amp;= MN_{(D-1)\times Q}(\Lambda_N, \Sigma, \Gamma_N).
\end{align}
</annotation></semantics></math> If Laplace approximation is too slow,
unstable (see below) or simply not needed, the default behavior of
<em>pibble</em> is to preform the above matrix calculations and produce
a single point estimate of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Σ</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Λ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
based on the posterior means of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Σ</mi><mo stretchy="false" form="prefix">|</mo><mi>η</mi><mo>,</mo><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">p(\Sigma | \eta, X)</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mi>Λ</mi><mo stretchy="false" form="prefix">|</mo><mi>Σ</mi><mo>,</mo><mi>η</mi><mo>,</mo><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(\Lambda | \Sigma, \eta, X)</annotation></semantics></math>.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-aitchison1986" class="csl-entry">
Aitchison, J. 1986. <em>The Statistical Analysis of Compositional
Data</em>. Book. Monographs on Statistics and Applied Probability.
London ; New York: Chapman; Hall.
</div>
<div id="ref-gevers2014" class="csl-entry">
Gevers, Dirk, Subra Kugathasan, Lee A Denson, Yoshiki Vázquez-Baeza,
Will Van Treuren, Boyu Ren, Emma Schwager, et al. 2014. <span>“The
Treatment-Naive Microbiome in New-Onset Crohn’s Disease.”</span>
<em>Cell Host &amp; Microbe</em> 15 (3): 382–92.
</div>
<div id="ref-silverman2018" class="csl-entry">
Silverman, Justin D, Heather Durand, Rachael J Bloom, Sayan Mukherjee,
and Lawrence A David. 2018. <span>“Dynamic Linear Models Guide Design
and Analysis of Microbiota Studies Within Artificial Human Guts.”</span>
<em>bioRxiv</em>. <a href="https://doi.org/10.1101/306597" class="external-link">https://doi.org/10.1101/306597</a>.
</div>
<div id="ref-silverman2019" class="csl-entry">
Silverman, Justin D., Kimberly Roche, Zachary C. Holmes, Lawrence A.
David, and Sayan Mukherjee. 2019. <span>“<span class="nocase">Bayesian
Multinomial Logistic Normal Models through Marginally Latent Matrix-T
Processes</span>.”</span> <em>arXiv e-Prints</em>, March,
arXiv:1903.11695. <a href="https://arxiv.org/abs/1903.11695" class="external-link">https://arxiv.org/abs/1903.11695</a>.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://www.justin-silverman.com/" class="external-link">Justin Silverman</a>, Michelle Nixon.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
