<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Joint Modeling  (e.g., Multiomics) with fido::Orthus ‚Ä¢ fido</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Joint Modeling  (e.g., Multiomics) with fido::Orthus">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fido</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introduction-to-fido.html">Introduction to fido::Pibble</a></li>
    <li><a class="dropdown-item" href="../articles/non-linear-models.html">Non-linear models with fido::basset</a></li>
    <li><a class="dropdown-item" href="../articles/orthus.html">Joint Modeling  (e.g., Multiomics) with fido::Orthus</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jsilve24/fido/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Joint Modeling  (e.g., Multiomics) with fido::Orthus</h1>
                        <h4 data-toc-skip class="author">Justin
Silverman</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jsilve24/fido/blob/master/vignettes/orthus.Rmd" class="external-link"><code>vignettes/orthus.Rmd</code></a></small>
      <div class="d-none name"><code>orthus.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jsilve24.github.io/fido/">fido</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">48482</span><span class="op">)</span></span></code></pre></div>
<p>If you have not already done so, I would read through the
<em>pibble</em> vignette before this one.</p>
<div class="section level2">
<h2 id="fidoorthus-the-model">fido::orthus, the model<a class="anchor" aria-label="anchor" href="#fidoorthus-the-model"></a>
</h2>
<p>fido can be used for jointly modeling multivariate count data and
multivariate Gaussian data. For example, this would be a reasonable
model to jointly model 16S microbiome data and metabolomics data
jointly. Because of the ‚Äútwo-headed‚Äù nature of this model, e.g., two
observed data-sets, I named this model <em>orthus</em>, <a href="https://en.Wikipedia.org/wiki/Orthrus" class="external-link">a two-headed dog and
brother of Cerberus in Greek Mythology</a>. The <em>orthus</em> model
can be written as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Y</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>‚àº</mo><mtext mathvariant="normal">Multinomial</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>œÄ</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>œÄ</mi><mi>j</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mi>œï</mi><mrow><mo>‚àí</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Œ∑</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="center" style="text-align: center"><msub><mi>Œ∑</mi><mi>j</mi></msub></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><msub><mi>Z</mi><mi>j</mi></msub></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>‚àº</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œõ</mi><mi>X</mi><mo>,</mo><mi>Œ£</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>Œõ</mi></mtd><mtd columnalign="left" style="text-align: left"><mo>‚àº</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œò</mi><mo>,</mo><mi>Œ£</mi><mo>,</mo><mi>Œì</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>Œ£</mi></mtd><mtd columnalign="left" style="text-align: left"><mo>‚àº</mo><msup><mi>W</mi><mrow><mo>‚àí</mo><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œû</mi><mo>,</mo><mi>œÖ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j \right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\begin{bmatrix}\eta_j \\ Z_j \end{bmatrix} &amp;\sim N(\Lambda X, \Sigma) \\
\Lambda &amp;\sim  N(\Theta, \Sigma, \Gamma) \\
\Sigma &amp;\sim W^{-1}(\Xi, \upsilon) 
\end{align}
</annotation></semantics></math></p>
<p>Note this looks nearly identical to the <em>pibble</em> model but we
have appended the second (Gaussian) dataset
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>)
onto
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ∑</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>.
In doing this, the definition of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œõ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
changes (it is now larger with the bottom rows dictating how the
covariates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
influence the second dataset). Similarly,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ£</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>
now is much larger and can be though of as
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Œ£</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="center" style="text-align: center"><msub><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œ∑</mi><mo>,</mo><mi>Œ∑</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mtd><mtd columnalign="center" style="text-align: center"><msub><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œ∑</mi><mo>,</mo><mi>Z</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><msub><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><mo>,</mo><mi>Œ∑</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mtd><mtd columnalign="center" style="text-align: center"><msub><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><mo>,</mo><mi>Z</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">
\Sigma = \begin{bmatrix} \Sigma_{(\eta, \eta)} &amp; \Sigma_{(\eta, Z)} \\
                          \Sigma_{(Z, \eta)} &amp; \Sigma_{(Z, Z)}\end{bmatrix}
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œ∑</mi><mo>,</mo><mi>Œ∑</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub><annotation encoding="application/x-tex">\Sigma_{(\eta, \eta)}</annotation></semantics></math>
describes the covariance between log-ratios (e.g., the covariance among
the multinomial categories in log-ratio space),
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><mo>,</mo><mi>Z</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub><annotation encoding="application/x-tex">\Sigma_{(Z, Z)}</annotation></semantics></math>
describes the covariance between the dimensions of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>
(e.g., between metabolites if Z is metabolomics data), and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œ∑</mi><mo>,</mo><mi>Z</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub><mo>=</mo><msubsup><mi>Œ£</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><mo>,</mo><mi>Œ∑</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>T</mi></msubsup></mrow><annotation encoding="application/x-tex">\Sigma_{(\eta, Z)} = \Sigma_{(Z, \eta)}^T</annotation></semantics></math>
represents the covariance between log-ratios and dimensions of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>
(e.g., between microbial taxa and metabolites). Similar to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ£</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œõ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>,
the parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œû</mi><annotation encoding="application/x-tex">\Xi</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œò</mi><annotation encoding="application/x-tex">\Theta</annotation></semantics></math>
undergo a similar expansion to accommodate the second dataset.</p>
</div>
<div class="section level2">
<h2 id="joint-modeling-of-microbial-16s-data-and-metabolomics">Joint modeling of Microbial 16S data and Metabolomics<a class="anchor" aria-label="anchor" href="#joint-modeling-of-microbial-16s-data-and-metabolomics"></a>
</h2>
<p>To demonstrate <em>orthus</em> I will perform a toy analysis on data
from <span class="citation">Kashyap et al. (2013)</span> and made
available by <span class="citation">Callahan et al. (2016)</span> as
part of their recently published microbiome data analysis workflow <span class="citation">(Callahan et al. 2016)</span>. I follow the data
preprocessing of <span class="citation">Callahan et al. (2016)</span> we
just don‚Äôt drop taxa but instead amalgamate those that don‚Äôt pass
filtering to a category called ‚Äúother‚Äù. I do this to maintain the proper
variance in the multinomial model.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metab_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/Kashyap2013"</span>, <span class="st">"metabolites.csv"</span>, package<span class="op">=</span><span class="st">"fido"</span><span class="op">)</span></span>
<span><span class="va">microbe_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/Kashyap2013"</span>, <span class="st">"microbe.rda"</span>, package<span class="op">=</span><span class="st">"fido"</span><span class="op">)</span></span>
<span><span class="va">metab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">metab_path</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">metab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">metab</span><span class="op">)</span></span>
<span><span class="va">microbe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html" class="external-link">get</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">microbe_path</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Preprocessing ##</span></span>
<span><span class="co"># Metabolite Preprocessing</span></span>
<span><span class="va">keep_ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">metab</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">3</span></span>
<span><span class="va">metab</span> <span class="op">&lt;-</span> <span class="va">metab</span><span class="op">[</span><span class="va">keep_ix</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 16S Preprocesing - plus some weirdness to rename amalgamated category to "other"</span></span>
<span><span class="va">keep_ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">4</span></span>
<span><span class="va">keep_ix</span> <span class="op">&lt;-</span> <span class="va">keep_ix</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html" class="external-link">otu_table</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">2</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">microbe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/merge_taxa-methods.html" class="external-link">merge_taxa</a></span><span class="op">(</span><span class="va">microbe</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">keep_ix</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">nms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span></span>
<span><span class="va">rnm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">keep_ix</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">nms</span><span class="op">[</span><span class="va">rnm</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"other"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">nms</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">nms</span>, <span class="va">rnm</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># bit of preprocessing </span></span>
<span><span class="va">metab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">metab</span><span class="op">)</span></span></code></pre></div>
<p>Now I am going to do just a bit of processing to get data into a
format for orthus. Note I have no extra metadata so we are just going to
use an intercept in our model at this time.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html" class="external-link">otu_table</a></span><span class="op">(</span><span class="va">microbe</span>, taxa_are_rows<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="va">metab</span> <span class="co">#(metabolites are rows)</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">microbe</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save dims for easy reference</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="va">P</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span></span>
<span><span class="va">Q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span></code></pre></div>
<p>Now I am going to set up the priors. My priors are going to be
similar to that of <em>pibble</em> but now we need to think about a
prior for the covariance among the metabolites and between the
metabolites and the log-ratios of the taxa. Remember, that priors must
be defined in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>L</mi><msub><mi>R</mi><mi>D</mi></msub></mrow><annotation encoding="application/x-tex">ALR_D</annotation></semantics></math>
(e.g., ALR with the reference being the D-th taxa; this may be changed
in the future to make specifying priors more user friendly).</p>
<p>I am going to form our prior for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ£</mi><annotation encoding="application/x-tex">\Sigma</annotation></semantics></math>
by specifying
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>œÖ</mi><annotation encoding="application/x-tex">\upsilon</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œû</mi><annotation encoding="application/x-tex">\Xi</annotation></semantics></math>.
I will specify that I have weak prior belief that the taxa are
independent in terms of their log absolute abundance. We can translate
this statement about covariance of log absolute abundance into a
statement about log-ratio covariance by pre- and post-multiplying by the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>L</mi><msub><mi>R</mi><mi>D</mi></msub></mrow><annotation encoding="application/x-tex">ALR_D</annotation></semantics></math>
contrast matrix (which I refer to as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mi>G</mi></mrow><annotation encoding="application/x-tex">GG</annotation></semantics></math>
below). Additionally, I believe that there is likely no substantial
covariance between the taxa and the metabolites and I assume the
metabolites are likely independent.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">upsilon</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">D</span><span class="op">-</span><span class="fl">1</span><span class="op">+</span><span class="va">P</span><span class="op">)</span><span class="op">+</span><span class="fl">10</span> <span class="co"># weak-ish prior on covariance over joint taxa and metabolites</span></span>
<span><span class="va">Xi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">D</span><span class="op">-</span><span class="fl">1</span><span class="op">+</span><span class="va">P</span><span class="op">)</span></span>
<span><span class="va">GG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">D</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">Xi</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="va">D</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="va">D</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">GG</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">GG</span><span class="op">)</span></span>
<span><span class="va">Xi</span> <span class="op">&lt;-</span> <span class="va">Xi</span> <span class="op">*</span> <span class="op">(</span><span class="va">upsilon</span><span class="op">-</span><span class="va">D</span><span class="op">-</span><span class="va">P</span><span class="op">)</span> <span class="co"># this scales Xi to have the proper mean we wanted</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">Xi</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://github.com/jsilve24/fido/raw/master/vignettes/orthus-prior-structure.png" width="50%" style="display: block; margin: auto;"></p>
<p>Note the structure of this prior, everything is independent but there
is a moderate positive covariance between the log-ratios based on their
shared definition in terms of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>D</mi><annotation encoding="application/x-tex">D</annotation></semantics></math>-th
taxa.</p>
<p>The other parts of the prior are less interesting. We are going to
state that our mean for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œõ</mi><annotation encoding="application/x-tex">\Lambda</annotation></semantics></math>
is centered about
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>ùüé</mn><annotation encoding="application/x-tex">\mathbf{0}</annotation></semantics></math>
and that the signal-to-noise ratio in the data is approximately 1 (this
later part is specified by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Œì</mi><mo>=</mo><mi>I</mi></mrow><annotation encoding="application/x-tex">\Gamma=I</annotation></semantics></math>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Gamma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Q</span><span class="op">)</span></span>
<span><span class="va">Theta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">D</span><span class="op">-</span><span class="fl">1</span><span class="op">+</span><span class="va">P</span>, <span class="va">Q</span><span class="op">)</span></span></code></pre></div>
<p>Finally I fit the model.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/orthus_fit.html">orthus</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">Z</span>, <span class="va">X</span>, Theta<span class="op">=</span><span class="va">Theta</span>, Gamma<span class="op">=</span><span class="va">Gamma</span>, Xi<span class="op">=</span><span class="va">Xi</span>, upsilon<span class="op">=</span><span class="va">upsilon</span>, n_samples<span class="op">=</span><span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<p>Next we are going to transform the log-ratios from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>L</mi><msub><mi>R</mi><mi>D</mi></msub></mrow><annotation encoding="application/x-tex">ALR_D</annotation></semantics></math>
to the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>L</mi><mi>R</mi></mrow><annotation encoding="application/x-tex">CLR</annotation></semantics></math>.
I have written all the transformation functions, <em>e.g.</em>,
<code>to_clr</code> etc‚Ä¶ to work on <code>orthusfit</code> objects in a
similar manner to how they work on <code>pibblefit</code> objects. For
<code>orthusfit</code> objects they only transform the log-ratio
components of parameters leaving the other parts of inferred model
parameters (<em>i.e.</em>, the parts associated with the metabolites)
untouched.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fido_transforms.html">to_clr</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="co">#&gt; orthusfit Object: </span></span>
<span><span class="co">#&gt;   Number of Samples:      12 </span></span>
<span><span class="co">#&gt;   Number of Categories:       114 </span></span>
<span><span class="co">#&gt;   Number of Zdimensions:  405 </span></span>
<span><span class="co">#&gt;   Number of Covariates:       1 </span></span>
<span><span class="co">#&gt;   Number of Posterior Samples:    1000 </span></span>
<span><span class="co">#&gt;   Contains Samples of Parameters:Eta  Lambda  Sigma</span></span>
<span><span class="co">#&gt;   Coordinate System:      clr</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="investigate-model-results">Investigate Model Results<a class="anchor" aria-label="anchor" href="#investigate-model-results"></a>
</h2>
<p>There are a ton of ways to visualize the inferred model. I could make
network diagrams relating taxa to taxa, taxa to metabolites and
metabolites to metabolites. I could look at a low dimensional
representation of joint covariance to create something very much akin to
canonical correlation analysis (CCA). I could look at how well the
metabolites predict the taxa and vice-versa. But for the sake of
simplicity I will do something much simpler. Here I am just going to
find a list of taxa metabolite covariances that the model is very
confident about.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First just look ath the cross-covariances fit by the model</span></span>
<span><span class="co"># (covariance between taxa in CLR coordinates and metabolites)</span></span>
<span><span class="co"># This requires that we extract the corner of Sigma. </span></span>
<span><span class="va">xcor</span> <span class="op">&lt;-</span> <span class="va">fit</span><span class="op">$</span><span class="va">Sigma</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">D</span>, <span class="va">D</span><span class="op">:</span><span class="op">(</span><span class="va">D</span><span class="op">-</span><span class="fl">1</span><span class="op">+</span><span class="va">P</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Initial preprocessing to speed up computation of posterior intervals</span></span>
<span><span class="co"># As there are a lot of cross-covariance terms we are going to first </span></span>
<span><span class="co"># weed down the list of things we have to look at by first pass </span></span>
<span><span class="co"># selecting only those taxa that have a large posterior mean for the covariance</span></span>
<span><span class="va">xcor.mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">xcor</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">mean</span><span class="op">)</span></span>
<span><span class="va">to.analyze</span> <span class="op">&lt;-</span> <span class="fu">fido</span><span class="fu">::</span><span class="fu"><a href="../reference/gather_array.html">gather_array</a></span><span class="op">(</span><span class="va">xcor.mean</span>, <span class="va">cov</span>, <span class="va">taxa</span>, <span class="va">metabolite</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">cov</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>,<span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>tm <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">taxa</span>, <span class="st">"_"</span>, <span class="va">metabolite</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset Covariance to those we are interested in and calculate posterior </span></span>
<span><span class="co"># confidence intervals. </span></span>
<span><span class="va">xcor.summary</span> <span class="op">&lt;-</span> <span class="fu">fido</span><span class="fu">::</span><span class="fu"><a href="../reference/gather_array.html">gather_array</a></span><span class="op">(</span><span class="va">xcor</span>, <span class="va">cov</span>, <span class="va">taxa</span>, <span class="va">metabolite</span>, <span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>tm<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">taxa</span>, <span class="st">"_"</span>, <span class="va">metabolite</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">tm</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">to.analyze</span><span class="op">$</span><span class="va">tm</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>taxa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">[</span><span class="va">taxa</span><span class="op">]</span>, metabolite <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">[</span><span class="va">metabolite</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">taxa</span>, <span class="va">metabolite</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">fido</span><span class="fu">:::</span><span class="fu"><a href="../reference/summarise_posterior.html">summarise_posterior</a></span><span class="op">(</span><span class="va">cov</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">taxa</span> <span class="op">!=</span> <span class="st">'other'</span><span class="op">)</span> <span class="co"># we don't care about these</span></span>
<span></span>
<span><span class="co"># Select those covariances where the model has high certainty (95%) that</span></span>
<span><span class="co"># the true covariance is not zero. </span></span>
<span><span class="va">xcor.summary</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">p2.5</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">p97.5</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 218 x 8</span></span>
<span><span class="co">#&gt; # Groups:   taxa [17]</span></span>
<span><span class="co">#&gt;    taxa  metabolite   p2.5   p25   p50  mean   p75  p97.5</span></span>
<span><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 722   206.0445922 -6.55 -3.96 -3.08 -3.33 -2.32 -1.51 </span></span>
<span><span class="co">#&gt;  2 7816  206.0445922 -5.63 -3.23 -2.45 -2.65 -1.85 -1.05 </span></span>
<span><span class="co">#&gt;  3 722   290.9298419 -5.18 -3.09 -2.42 -2.62 -1.85 -1.23 </span></span>
<span><span class="co">#&gt;  4 18182 380.1846197 -5.20 -3.07 -2.38 -2.55 -1.83 -1.03 </span></span>
<span><span class="co">#&gt;  5 722   181.4504354 -5.19 -3.04 -2.36 -2.55 -1.80 -1.13 </span></span>
<span><span class="co">#&gt;  6 722   177.0565368 -4.98 -3.01 -2.30 -2.50 -1.78 -1.08 </span></span>
<span><span class="co">#&gt;  7 722   180.072273  -5.03 -3.06 -2.30 -2.49 -1.71 -0.986</span></span>
<span><span class="co">#&gt;  8 19517 380.1846197 -5.15 -3.06 -2.33 -2.49 -1.71 -0.952</span></span>
<span><span class="co">#&gt;  9 2943  380.1846197 -4.89 -2.97 -2.34 -2.49 -1.83 -1.03 </span></span>
<span><span class="co">#&gt; 10 722   176.0343919 -4.93 -2.95 -2.25 -2.48 -1.79 -1.09 </span></span>
<span><span class="co">#&gt; # ... with 208 more rows</span></span></code></pre></div>
<p>So it looks there there are a few hundred covariances that we can be
fairly confident about.</p>
</div>
<div class="section level2">
<h2 id="qualifications-and-caution">Qualifications and Caution<a class="anchor" aria-label="anchor" href="#qualifications-and-caution"></a>
</h2>
<p>Please note, I performed this analysis to demonstrate the use of
<em>orthus</em> which is a model that I have been repeatedly asked for.
I think its a cool model and could be quite useful in the right
circumstances. But I would like to point out a few philosophical points
about the analysis I performed above.</p>
<p>First, I performed this analysis just to demonstrate <em>orthus</em>.
I really don‚Äôt know the data showcased here. What is metabolite
<code>206.0445922</code>? I have no idea. For some reason this is how
the metabolites in that dataset were named. For the same reason I have
left the taxa indexed by sequence variant number.</p>
<p>Second (and more important), identifying relationships between taxa
and metabolites (or between any two high-dimensional multivariate
data-sets) is really difficult! Here we are looking at just 114 taxa and
405 but this leads to 46170 possible covariances and here we only have
12 samples! Yes <em>orthus</em> is a Bayesian model, and Yes, Bayesian
models can be quite useful when there are more parameters than samples,
but there is a limit of reasonability. Really, Bayesian models are great
when you can perfectly capture your prior beliefs with your prior. But
how often can that really be done perfectly? As such I would caution
users, to use <em>orthus</em> carefully. Consider which metabolites and
taxa you really care about and if you can, isolate your analyses to
those.</p>
<p>Alright, that‚Äôs probably enough philosophizing for an R package
Vignette. I hope you enjoy <em>orthus</em>.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-callahan2016" class="csl-entry">
Callahan, Ben J, Kris Sankaran, Julia A Fukuyama, Paul J McMurdie, and
Susan P Holmes. 2016. <span>‚ÄúBioconductor Workflow for Microbiome Data
Analysis: From Raw Reads to Community Analyses.‚Äù</span>
<em>F1000Research</em> 5.
</div>
<div id="ref-kashyap2013" class="csl-entry">
Kashyap, Purna C, Angela Marcobal, Luke K Ursell, Samuel A Smits, Erica
D Sonnenburg, Elizabeth K Costello, Steven K Higginbottom, et al. 2013.
<span>‚ÄúGenetically Dictated Change in Host Mucus Carbohydrate Landscape
Exerts a Diet-Dependent Effect on the Gut Microbiota.‚Äù</span>
<em>Proceedings of the National Academy of Sciences</em> 110 (42):
17059‚Äì64.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://www.justin-silverman.com/" class="external-link">Justin Silverman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
